{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3b80dc4c-2d4a-4819-9806-857415b8655c",
   "metadata": {},
   "source": [
    "# import packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9963c61c-3057-4c1b-a6cc-a5fc8e7ee7f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(reticulate)\n",
    "use_condaenv(\"mar25\", required = TRUE)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4baf4578-5710-4a4e-879b-f233d0c217bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(Signac)\n",
    "library(Seurat)\n",
    "library(data.table)\n",
    "library(ggplot2)\n",
    "library(tidyverse)\n",
    "library(readxl)\n",
    "library(patchwork)\n",
    "library(ComplexHeatmap)\n",
    "library(patchwork)\n",
    "library(ggpubr)\n",
    "library(GenomicRanges) \n",
    "library(EnsDb.Mmusculus.v79)\n",
    "library(DT)\n",
    "library(future)\n",
    "library(anndata)\n",
    "\n",
    "plan(\"multicore\", workers = 4)\n",
    "\n",
    "options(future.globals.maxSize = 50000 * 1024^2) # for 50 Gb RAM\n",
    "\n",
    "#base_dir = '/research_jude/rgs01_jude/groups/jxugrp/home/common/Lab_Members/WenhuoHu/collab_JinWang/'\n",
    "base_dir = '/research_jude/rgs01_jude/groups/jxugrp/home/common/Lab_Members/WenhuoHu/repos/niche/'\n",
    "setwd(base_dir)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8dd357b3-ed6d-4a34-a1e0-5dd9bdb68ba2",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "# import design table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e12cc2f-d835-4085-a913-16868c85f975",
   "metadata": {},
   "outputs": [],
   "source": [
    "design = fread(file = 'data/dsn_dt_base_v4.tsv')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36555629-780d-457a-9741-20156ca77a0e",
   "metadata": {},
   "outputs": [],
   "source": [
    "ma9_msc_samples = c('M2', 'M3', 'M4', 'M4_m10', 'ctrl_m4w', 'ctrl_m6w', 'ctrl_m8w')\n",
    "\n",
    "msc_design = design[sample_name %in% ma9_msc_samples,] \n",
    "msc_design\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f591416-87fb-432e-a39e-c736b07c9b3e",
   "metadata": {},
   "source": [
    "# import MSC RNAseq analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "00ba7231-59aa-40f6-afc5-0868487532b0",
   "metadata": {},
   "source": [
    "## import all data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e4726eb-8aa7-40e8-b492-e271aa726dfd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# h5 files were generated by cellranger\n",
    "h5_list = lapply(msc_design$h5file, Read10X_h5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb8f0d4e-c8ea-4c38-976e-b33cd6818176",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create seurat object using the gene expression data\n",
    "rna_seu_list = lapply(1:nrow(msc_design), function(ii){\n",
    "    h5 = h5_list[[ii]]\n",
    "    sampleID2 = msc_design$sampleID2[ii]\n",
    "    oo = CreateSeuratObject(counts = h5$`Gene Expression`, assay = 'RNA', project = sampleID2)\n",
    "    RenameCells(object = oo, new.names = paste0(sampleID2, '__', Cells(oo)))\n",
    "})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8495a1eb-8cba-4f6a-a1b9-337ddd37978f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create seurat object using the atac sequencing data\n",
    "ata_seu_list = lapply(1:nrow(msc_design), function(ii){\n",
    "    h5 = h5_list[[ii]]\n",
    "    sampleID2 = msc_design$sampleID2[ii]\n",
    "    oo = CreateSeuratObject(counts = h5$`Peaks`, assay = 'ATAC', project = sampleID2)\n",
    "    RenameCells(object = oo, new.names = paste0(sampleID2, '__', Cells(oo)))\n",
    "})\n",
    "# CreateAssay5Object(counts = pbmc.counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3eed7604-3a81-458b-8a2e-39c89443ef12",
   "metadata": {},
   "outputs": [],
   "source": [
    "# merge all samples\n",
    "rna_seu  = merge(rna_seu_list[[1]], rna_seu_list[2:length(rna_seu_list)], labels = msc_design$sampleID2)\n",
    "ata_seu  = merge(ata_seu_list[[1]], ata_seu_list[2:length(ata_seu_list)], labels = msc_design$sampleID2)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6004493a-b4f8-40ae-ac7c-c345842da6bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_seu\n",
    "ata_seu"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f8ded57-1155-4c0a-b3c0-bccd49755cae",
   "metadata": {},
   "outputs": [],
   "source": [
    "# checking cell meta data\n",
    "tmp = data.table(id = Cells(rna_seu))\n",
    "tmp[, sampleID := sub('__.*', '', id)] \n",
    "tmp[, cellID_n := sub('.*__', '', id)] \n",
    "tmp[, cellID   := sub('_.*', '', cellID_n)] \n",
    "tmp[, n := sub('.*_', '', cellID_n)] \n",
    "head(tmp)\n",
    "table(tmp$sampleID, tmp$n)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "02045197-6d1d-454a-ac41-1617342fb2d2",
   "metadata": {},
   "source": [
    "## select RNAseq based on previous analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "47c63a79-c00b-4201-9124-24204c3b8837",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import previous data only for MSC cell IDs \n",
    "#  ~ 3900 cells\n",
    "msc_rna_2 = read_h5ad(\"data/msc_rna_2_mar12.h5ad\") "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "297a7f3b-5a3f-47a5-8d16-7b8509c1431b",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_ids = as.data.table(msc_rna_2$obs, keep.rownames = T)\n",
    "rna_ids[, id := rn] \n",
    "head(rna_ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54120ac2-7ded-4407-8943-92aa638dd439",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_ids[, cell_ID := sub('_.*', '', id) ]\n",
    "rna_ids[, sample_ID := sub('.*_', '', id) ]\n",
    "table(rna_ids$sample_ID)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eab6c6f2-2748-4efa-a0f5-bc6e51e09e63",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_ids[, cell_ID := sub('_.*', '', id) ]\n",
    "rna_ids[, sample_ID := sub('.*_', '', id) ]\n",
    "rna_ids[, sample_ID2 := sample_ID]\n",
    "rna_ids[sample_ID == 'm10', sample_ID2 := 'M4']\n",
    "\n",
    "rna_ids[sample_ID2 == 'M2', sampleID := 'MA9_MSC_wk4']\n",
    "rna_ids[sample_ID2 == 'M3', sampleID := 'MA9_MSC_wk6']\n",
    "rna_ids[sample_ID2 == 'M4', sampleID := 'MA9_MSC_wk8']\n",
    "rna_ids[sample_ID2 == 'm4w', sampleID := 'Ctrl_MSC_wk4']\n",
    "rna_ids[sample_ID2 == 'm6w', sampleID := 'Ctrl_MSC_wk6']\n",
    "rna_ids[sample_ID2 == 'm8w', sampleID := 'Ctrl_MSC_wk8']\n",
    "\n",
    "rna_ids[sample_ID == 'M2',  nn := 1 ]\n",
    "rna_ids[sample_ID == 'M3',  nn := 2 ]\n",
    "rna_ids[sample_ID == 'M4',  nn := 3 ]\n",
    "rna_ids[sample_ID == 'm10', nn := 4 ]\n",
    "rna_ids[sample_ID == 'm4w', nn := 5 ]\n",
    "rna_ids[sample_ID == 'm6w', nn := 6 ]\n",
    "rna_ids[sample_ID == 'm8w', nn := 7 ]\n",
    "\n",
    "rna_ids[, seuratID := paste0(sampleID, '__', cell_ID, '_', nn)]\n",
    "head(rna_ids)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55947e9a-080f-4326-935b-64c64f68d07b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# subset cells based on these IDs\n",
    "rna_seu_2 = subset(rna_seu, cells = Cells(rna_seu)[Cells(rna_seu) %in% rna_ids$seuratID])\n",
    "\n",
    "# atac seq will be processed by importing raw data below\n",
    "#ata_seu_2 = subset(ata_seu, cells = Cells(ata_seu)[Cells(ata_seu) %in% rna_ids$seuratID])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e16697c4-f1a8-4681-b9c2-08f55c9069ea",
   "metadata": {},
   "source": [
    "## preprocess for selected RNA "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f75c030f-c367-4bfc-a1df-4b6c08512147",
   "metadata": {},
   "outputs": [],
   "source": [
    "seurat_obj <- FindVariableFeatures(seurat_obj, selection.method = \"vst\", nfeatures = 2000)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e8e9677-7511-491f-aa1c-d46a8bb3d251",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_seu_2 <- SCTransform(rna_seu_2, verbose = FALSE, return.only.var.genes = F)\n",
    "rna_seu_2 <- RunPCA(rna_seu_2, verbose = FALSE)\n",
    "rna_seu_2 <- RunUMAP(rna_seu_2, dims = 1:30, verbose = FALSE)\n",
    "\n",
    "rna_seu_2 <- FindNeighbors(rna_seu_2, dims = 1:30, verbose = FALSE)\n",
    "rna_seu_2 <- FindClusters(rna_seu_2, verbose = FALSE, resolution = .2)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2682246d-33a7-4ddf-8007-cf3f27bed818",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 10, repr.plot.height = 5, repr.plot.res = 200)\n",
    "DimPlot(rna_seu_2, group.by = c(\"orig.ident\", \"seurat_clusters\"), label = TRUE) + ggtitle(\"RNA\") \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11335d61-3cde-406c-90fa-03eb06cd5be6",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_seu_2$umap1 = as.data.frame(Embeddings(rna_seu_2, 'umap'))$umap_1\n",
    "rna_seu_2$umap2 = as.data.frame(Embeddings(rna_seu_2, 'umap'))$umap_2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0fa8f6e2-1e33-4bea-9789-9dd30d81dad7",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_seu_3 <- subset(rna_seu_2, idents = c(0, 1, 2))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c800f5ef-dd58-4e05-b90c-7a54f9451199",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_seu_3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e1739d9-26b8-4bf4-afc1-6e3e9ed9c28a",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 10, repr.plot.height = 5, repr.plot.res = 200)\n",
    "DimPlot(rna_seu_3, group.by = c(\"orig.ident\", \"seurat_clusters\"), label = TRUE) + ggtitle(\"RNA\") \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21a6e72e-4484-4e91-8599-1476d1ae0990",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add genome information \n",
    "# annotations <- GetGRangesFromEnsDb(ensdb = EnsDb.Mmusculus.v79)\n",
    "# seqlevelsStyle(annotations) <- \"UCSC\"\n",
    "# genome(annotations) <- \"mm10\"\n",
    "# Annotation(ata_seu_2) <- annotations \n",
    "# ata_seu_3 = subset(ata_seu_2, cells = Cells(rna_seu_3)) \n",
    "# We exclude the first dimension as this is typically correlated with sequencing depth\n",
    "# DefaultAssay(ata_seu_3) <- \"ATAC\"\n",
    "# ata_seu_3 <- RunTFIDF(ata_seu_3, assay = \"ATAC\")\n",
    "# ata_seu_3 <- FindTopFeatures(ata_seu_3, min.cutoff = \"q0\")\n",
    "# ata_seu_3 <- RunSVD(ata_seu_3)\n",
    "# ata_seu_3 <- RunUMAP(ata_seu_3, reduction = \"lsi\", dims = 2:30, reduction.name = \"umap.atac\", reduction.key = \"atacUMAP_\", assay = \"ATAC\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "110fc872-1edc-42c1-8a99-a79c99bc2faa",
   "metadata": {},
   "outputs": [],
   "source": [
    "ata_seu_3 = subset(ata_seu_2, cells = Cells(rna_seu_3)) \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc75f237-8b0c-40ca-b54f-7db8ce730ba2",
   "metadata": {},
   "source": [
    "# exported files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c364dd60-fbeb-4832-8817-c96bcc074270",
   "metadata": {},
   "outputs": [],
   "source": [
    "# saveRDS(ata_seu, file = 'data/ata_seu.rds') \n",
    "# saveRDS(ata_seu_5, file = 'data/ata_seu_5.rds') \n",
    "# saveRDS(cds, file = 'data/cds_monocle3.rds') \n",
    "# saveRDS(oo_msc_5, file = 'data/oo_msc_5.rds')\n",
    "# saveRDS(oo_msc_5_peaks, file = 'data/oo_msc_5_peaks.rds')\n",
    "# saveRDS(oo_msc_5_peaks_4clusters, file = 'data/oo_msc_5_peaks_4clusters.rds')\n",
    "# saveRDS(rna_seu, file = 'data/rna_seu.rds') \n",
    "# saveRDS(rna_seu_5, file = 'data/rna_seu_5.rds') \n",
    "# saveRDS(rna_seu_5_markers, file = 'data/rna_seu_5_markers.rds')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "851154a7-ac3c-4b49-beb3-67dc8664a53b",
   "metadata": {},
   "source": [
    "# import ATACSeq "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8603f1ef-93bf-43ec-a0e4-2a9dececee7a",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## import raw data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0408b5b2-5f41-45f7-9159-010e56800ecb",
   "metadata": {},
   "outputs": [],
   "source": [
    "atac_peak_list = lapply(msc_design$peak_file, function(fname){\n",
    "    tmp = read.table(fname, col.names = c(\"chr\", \"start\", \"end\"))\n",
    "    makeGRangesFromDataFrame(tmp)\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9299cfe9-b475-4b6a-a032-25c46da941c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "atac_peaks <- Signac::reduce(c(\n",
    "    atac_peak_list[[1]], atac_peak_list[[2]], atac_peak_list[[3]], atac_peak_list[[4]],\n",
    "    atac_peak_list[[5]], atac_peak_list[[6]], atac_peak_list[[7]]))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3669f3d3-710f-4294-bd13-b09631635ee4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter out bad peaks based on length\n",
    "peakwidths <- width(atac_peaks)\n",
    "atac_peaks <- atac_peaks[peakwidths  < 10000 & peakwidths > 20]\n",
    "atac_peaks\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a70026e1-6fb3-4e85-a455-a79a8f7f15f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "metrics_list = lapply(msc_design$metric_file, function(fname){\n",
    "    tmp <- read.table(fname, stringsAsFactors = FALSE, sep = \",\", header = TRUE, row.names = 1 )\n",
    "    tmp[-1, ]\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c0496cf-3986-43df-b8c1-93bf545968fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "summary(metrics_list[[1]]$atac_raw_reads)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eebedac4-e3e7-4a4a-b987-a4ec86d973f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "dim(metrics_list[[1]][metrics_list[[1]]$atac_raw_reads > 1500, ])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36193db2-b772-45b5-a57a-c256a8124a59",
   "metadata": {},
   "outputs": [],
   "source": [
    "mtrics_filtered_list = lapply(metrics_list, function(tmp){tmp[tmp$atac_raw_reads > 1000, ]})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "696d26ce-5ccb-481e-b810-7bfc59ffc88b",
   "metadata": {},
   "outputs": [],
   "source": [
    "lapply(mtrics_filtered_list, nrow)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0af83b3-5e27-4634-a3f0-2edc7f4b4bb7",
   "metadata": {},
   "outputs": [],
   "source": [
    "msc_frag_list = lapply(1:length(mtrics_filtered_list), function(ii){ \n",
    "    fname = msc_design[ii, fragment_file]\n",
    "    cells = mtrics_filtered_list[[ii]][, 'gex_barcode']\n",
    "    CreateFragmentObject( path = fname, cells = cells) }\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23d98011-c54d-4483-bb41-f696c0e3794b",
   "metadata": {},
   "outputs": [],
   "source": [
    "feature_counts_list = lapply(1:length(msc_frag_list), function(ii){\n",
    "    frag  = msc_frag_list[[ii]]\n",
    "    cells = mtrics_filtered_list[[ii]][, 'gex_barcode']\n",
    "    FeatureMatrix( fragments = frag, features = atac_peaks, cells = cells)}  )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75f416f6-6aeb-4a45-b773-89b82590677d",
   "metadata": {},
   "outputs": [],
   "source": [
    "atac_assay_list = lapply(1:length(feature_counts_list), function(ii){\n",
    "    CreateChromatinAssay(feature_counts_list[[ii]], fragments = msc_frag_list[[ii]])\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e9f065d-dac2-42f7-ba5b-e34b117e368d",
   "metadata": {},
   "outputs": [],
   "source": [
    "atac_oo_list = lapply(1:length(atac_assay_list), function(ii){\n",
    "    oo = CreateSeuratObject(atac_assay_list[[ii]], assay = 'ATAC', meta.data = msc_frag_list[[ii]])\n",
    "    oo$dataset = msc_design$sample_name\n",
    "    oo\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f10a7eef-b432-4f06-8607-2dcd1b23b321",
   "metadata": {},
   "outputs": [],
   "source": [
    "msc_design$sampleID2 = sub('r2', '', msc_design$sampleID)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a67a3249-0ad1-4199-93f4-0a06c7c8fd19",
   "metadata": {},
   "outputs": [],
   "source": [
    "msc_design$sampleID2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4491559-7a2c-4b77-90f2-57f8e2a27611",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## merge ATAC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f461b64-d729-47e8-b576-82e58040f891",
   "metadata": {},
   "outputs": [],
   "source": [
    "oo <- merge(\n",
    "  x = atac_oo_list[[1]],\n",
    "  y = atac_oo_list[2:length(atac_oo_list)],\n",
    "  add.cell.ids = paste0(msc_design$sampleID2, '_')\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "736559da-8679-4e2d-8d26-179ddec98ad9",
   "metadata": {},
   "outputs": [],
   "source": [
    "oo[[\"ATAC\"]]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "925bb233-6ad8-4a9c-92da-cfc07bcd4e53",
   "metadata": {},
   "source": [
    "## filter MSC ATAC sequencing cells based on RNASeq selected cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b74f1a2-a07a-46f3-9739-e0febac81734",
   "metadata": {},
   "outputs": [],
   "source": [
    "msc_dt = data.table(id = names(oo$dataset))\n",
    "msc_dt[, sample_ID := sub('__.*', '', id) ]\n",
    "msc_dt[, cell_ID := sub('.*__(.*)_.*', '\\\\1', id)]\n",
    "head(msc_dt) \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e03a65a-5929-4398-aa8e-543815269923",
   "metadata": {},
   "outputs": [],
   "source": [
    "table(msc_dt$sample_ID)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9cbbef3-77c1-457a-ba40-eedb72213064",
   "metadata": {},
   "outputs": [],
   "source": [
    "table(rna_ids$sample_ID2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17d1bd5d-7ec8-42e9-b443-104a686ce740",
   "metadata": {},
   "outputs": [],
   "source": [
    "length(intersect(msc_dt$id, rna_ids$seuratID))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f25cf45-1f93-4e56-a5c7-024cee48c9e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "msc_dt[, ov := F]\n",
    "msc_dt[id %in% rna_ids$seuratID, ov := T]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed6c9a1b-3cd8-4f97-b13b-411be1947596",
   "metadata": {},
   "outputs": [],
   "source": [
    "head(msc_dt[ov == T, ])\n",
    "nrow(msc_dt[ov == T, ])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c569b996-b98a-416d-bdca-59907374901d",
   "metadata": {},
   "outputs": [],
   "source": [
    "oo@meta.data$ov = msc_dt$ov"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e5544d4-d291-49e4-bf5f-ef1ceee8f34d",
   "metadata": {},
   "outputs": [],
   "source": [
    "oo_msc = subset(oo, ov == T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5bc26c74-507f-4549-b07b-ac755c84217f",
   "metadata": {},
   "outputs": [],
   "source": [
    "oo_msc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75404313-3723-4849-ad06-9c91dee15d80",
   "metadata": {},
   "outputs": [],
   "source": [
    "oo_msc <- RunTFIDF(oo_msc)\n",
    "oo_msc <- FindTopFeatures(oo_msc, min.cutoff = 20)\n",
    "oo_msc <- RunSVD(oo_msc)\n",
    "oo_msc <- RunUMAP(oo_msc, dims = 2:50, reduction = 'lsi')\n",
    "set.seed(1234)\n",
    "oo_msc <- FindNeighbors(oo_msc, dims = 2:50, reduction = \"lsi\")\n",
    "set.seed(1234)\n",
    "oo_msc <- FindClusters(oo_msc, resolution = 0.5)  # Adjust resolution as needed\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f6ea3a6-34ac-4caf-b159-f3e90f19037b",
   "metadata": {},
   "outputs": [],
   "source": [
    "DimPlot(oo_msc, reduction = \"umap\", label = TRUE)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c46afdf-13f6-4a6f-b612-cb939adf991d",
   "metadata": {},
   "outputs": [],
   "source": [
    "oo_msc@meta.data$umap1 = as.data.frame(Embeddings(oo_msc, reduction = 'umap'))$umap_1\n",
    "oo_msc@meta.data$umap2 = as.data.frame(Embeddings(oo_msc, reduction = 'umap'))$umap_2\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5cd51e00-d552-418f-ad48-b4df611b51cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "rm <- subset(oo_msc, umap1 > 4 & umap2 < -3)\n",
    "oo_msc_2 = subset(oo_msc, cells = setdiff(Cells(oo_msc), Cells(rm)))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0975c310-937d-4e92-bef6-203c6966a034",
   "metadata": {},
   "outputs": [],
   "source": [
    "oo_msc_2 <- RunTFIDF(oo_msc_2)\n",
    "oo_msc_2 <- FindTopFeatures(oo_msc_2, min.cutoff = 20)\n",
    "oo_msc_2 <- RunSVD(oo_msc_2)\n",
    "oo_msc_2 <- RunUMAP(oo_msc_2, dims = 2:50, reduction = 'lsi')\n",
    "set.seed(1234)\n",
    "oo_msc_2 <- FindNeighbors(object = oo_msc_2, reduction = 'lsi', dims = 2:30)\n",
    "set.seed(1234)\n",
    "oo_msc_2 <- FindClusters(object = oo_msc_2, verbose = FALSE, algorithm = 3, resolution = .1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87995e3b-5f7b-42f4-81ce-b9b9f322521f",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 10, repr.plot.height = 5, repr.plot.res = 200)\n",
    "DimPlot(rna_seu_2, group.by = c(\"orig.ident\", \"seurat_clusters\"), label = TRUE) + ggtitle(\"RNA\") \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b350a3f-bc95-4af4-b31c-f89802546779",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_seu_2$umap1 = as.data.frame(Embeddings(rna_seu_2, 'umap'))$umap_1\n",
    "rna_seu_2$umap2 = as.data.frame(Embeddings(rna_seu_2, 'umap'))$umap_2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2b02433-6ce4-45d9-aa25-eee15703a3f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_seu_3 = subset(rna_seu_2, subset =  umap1 > -5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33e162b6-2c52-415c-946b-8851fdf04829",
   "metadata": {},
   "outputs": [],
   "source": [
    "oo_msc_3 = subset(oo_msc_2, cells = Cells(rna_seu_3)) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "507e787d-07c7-4040-bca7-c42f6d3b74da",
   "metadata": {},
   "outputs": [],
   "source": [
    "ov_cells = intersect(Cells(oo_msc_2), Cells(rna_seu_2))\n",
    "length(ov_cells)\n",
    "\n",
    "oo_msc_3 = subset(oo_msc_2, cells = ov_cells) \n",
    "rna_seu_3 = subset(rna_seu_2, cells = ov_cells) \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ce14d08-35c3-4705-9a48-59e54d1a017f",
   "metadata": {},
   "outputs": [],
   "source": [
    "DefaultAssay(oo_msc_3) = 'ATAC'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92415f72-5ca0-403e-8055-9e7da9347615",
   "metadata": {},
   "outputs": [],
   "source": [
    "oo_msc_3 <- RunTFIDF(oo_msc_3)\n",
    "oo_msc_3 <- FindTopFeatures(oo_msc_3, min.cutoff = 20)\n",
    "oo_msc_3 <- RunSVD(oo_msc_3)\n",
    "oo_msc_3 <- RunUMAP(oo_msc_3, dims = 2:50, reduction = 'lsi')\n",
    "set.seed(1234)\n",
    "oo_msc_3 <- FindNeighbors(object = oo_msc_3, reduction = 'lsi', dims = 2:30)\n",
    "set.seed(1234)\n",
    "oo_msc_3 <- FindClusters(object = oo_msc_3, verbose = FALSE, algorithm = 3, resolution = .2)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1bad6606-ba9d-417d-8dac-b2352bd6e686",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 5, repr.plot.height = 4, repr.plot.res = 300)\n",
    "DimPlot(oo_msc_3, group.by = c('seurat_clusters'), pt.size = 0.1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6580d82c-bebc-41cf-8d29-c844ebef8773",
   "metadata": {},
   "outputs": [],
   "source": [
    "oo_msc_4 = subset(oo_msc_3, subset = seurat_clusters %in% c(0,1,2,3,4))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6bc6d6b2-af3e-466b-99ef-fef2a7e37d5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "oo_msc_4 <- RunTFIDF(oo_msc_4)\n",
    "oo_msc_4 <- FindTopFeatures(oo_msc_4, min.cutoff = 20)\n",
    "oo_msc_4 <- RunSVD(oo_msc_4)\n",
    "oo_msc_4 <- RunUMAP(oo_msc_4, dims = 2:50, reduction = 'lsi')\n",
    "set.seed(1234)\n",
    "oo_msc_4 <- FindNeighbors(object = oo_msc_4, reduction = 'lsi', dims = 2:30)\n",
    "set.seed(1234)\n",
    "oo_msc_4 <- FindClusters(object = oo_msc_4, verbose = FALSE, algorithm = 3, resolution = .2)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "204b337b-e4b7-4d34-9810-6e57bd924a14",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 5, repr.plot.height = 4, repr.plot.res = 300)\n",
    "DimPlot(oo_msc_4, group.by = c('seurat_clusters'), pt.size = 0.1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a7e7d3da-b13c-4ed6-9171-8cc02feb794a",
   "metadata": {},
   "outputs": [],
   "source": [
    "set.seed(1234)\n",
    "rna_seu_3 <- FindClusters(object = rna_seu_3, verbose = FALSE, algorithm = 3, resolution = .2)\n",
    "\n",
    "options(repr.plot.width = 10, repr.plot.height = 4, repr.plot.res = 300)\n",
    "DimPlot(rna_seu_3, group.by = c('orig.ident', 'seurat_clusters'), pt.size = 0.1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a7178b4-2fe8-4c5b-855e-5649556e0ccd",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "options(repr.plot.width = 5, repr.plot.height = 4, repr.plot.res = 300)\n",
    "DimPlot(rna_seu_3, group.by = c('seurat_clusters'), pt.size = 0.1) + geom_vline(xintercept = -3.4, color = 'grey70')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c18db4f9-21dc-4b01-9e24-3f7a87bc6db0",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_seu_3@meta.data$umap1 = as.data.frame(Embeddings(rna_seu_3, reduction = 'umap'))$umap_1\n",
    "rna_seu_3@meta.data$umap2 = as.data.frame(Embeddings(rna_seu_3, reduction = 'umap'))$umap_2\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "411d34d8-ec7c-4afa-9e37-75eeb4891cf6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#rna_seu_4 = subset(rna_seu_3, subset = seurat_clusters %in% c(0, 1, 2))\n",
    "rna_seu_4 = subset(rna_seu_3, subset = umap1 > -3.4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c4e40faa-3174-4979-9ac9-e357e02f1721",
   "metadata": {},
   "outputs": [],
   "source": [
    "ov_cells_4 = intersect(Cells(oo_msc_4), Cells(rna_seu_4))\n",
    "oo_msc_5 = subset(oo_msc_4, cells = ov_cells_4)\n",
    "rna_seu_5 = subset(rna_seu_4, cells = ov_cells_4)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3833e14-92e7-4d1e-8dbe-880c12da3d36",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 10, repr.plot.height = 4, repr.plot.res = 300)\n",
    "DimPlot(oo_msc_5, group.by = c('sampleID', 'seurat_clusters'), pt.size = 0.1)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "48345d98-547f-4e18-a5d5-2da3086deeba",
   "metadata": {},
   "source": [
    "# re-organize RNASeq datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36275f08-1f32-44da-97cd-5696094d8c16",
   "metadata": {},
   "outputs": [],
   "source": [
    "# to get consistent clusters, an better version were exported from previous run\n",
    "rna_seu_5 = readRDS('data/rna_seu_5.rds')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1aa4e9cc-8c4e-445a-a09b-163690685402",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_seu_5@meta.data$umap1 = as.data.frame(Embeddings(rna_seu_5, reduction = 'umap'))$umap_1\n",
    "rna_seu_5@meta.data$umap2 = as.data.frame(Embeddings(rna_seu_5, reduction = 'umap'))$umap_2\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6df7bf44-6935-4f56-8fde-2665a838f9eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 10, repr.plot.height = 4, repr.plot.res = 200)\n",
    "DimPlot(rna_seu_5, label = TRUE, group.by = c('orig.ident', 'seurat_clusters_adj'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1afd030b-f2a7-4ce6-8d74-37bf2255fe91",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_seu_5$seurat_clusters[rna_seu_5$umap2 < 0.5 & rna_seu_5$seurat_clusters == '5'] = '0' \n",
    "table(rna_seu_5$seurat_clusters)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15b11367-b9fd-46e6-ad41-f7eb1de6ed36",
   "metadata": {},
   "outputs": [],
   "source": [
    "# re-organize RNASeq cluster names\n",
    "rna_seu_5$seurat_clusters_adj = as.character(rna_seu_5$seurat_clusters)\n",
    "rna_seu_5$seurat_clusters_adj[rna_seu_5$seurat_clusters == '1'] = '1'\n",
    "rna_seu_5$seurat_clusters_adj[rna_seu_5$seurat_clusters == '6'] = '2'\n",
    "rna_seu_5$seurat_clusters_adj[rna_seu_5$seurat_clusters == '3'] = '3'\n",
    "rna_seu_5$seurat_clusters_adj[rna_seu_5$seurat_clusters == '0'] = '4'\n",
    "rna_seu_5$seurat_clusters_adj[rna_seu_5$seurat_clusters == '5'] = '5'\n",
    "rna_seu_5$seurat_clusters_adj[rna_seu_5$seurat_clusters == '4'] = '6'\n",
    "rna_seu_5$seurat_clusters_adj[rna_seu_5$seurat_clusters == '2'] = '7'\n",
    "rna_seu_5$seurat_clusters_adj = factor(rna_seu_5$seurat_clusters_adj , levels = c('1', '2', '3', '4', '5', '6', '7'))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "96060f1d-0379-44af-83d0-873956bcb753",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 10, repr.plot.height = 4, repr.plot.res = 200)\n",
    "DimPlot(rna_seu_5, label = TRUE, group.by = c('orig.ident', 'seurat_clusters_adj'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2945022a-b7a5-47d6-a137-86775b1bb678",
   "metadata": {},
   "outputs": [],
   "source": [
    "table(rna_seu_5$seurat_clusters_adj)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6cccc35e-ca91-4111-9819-0b4503c0c6c6",
   "metadata": {},
   "source": [
    "# re-organize ATAC-Seq data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bcbdaea9-bf9d-445d-bf22-3e3a202c595b",
   "metadata": {},
   "outputs": [],
   "source": [
    "oo_msc_5 = readRDS(file = 'data/oo_msc_5.rds')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57f6f4c0-5e2d-480b-a69a-1b6c45683f04",
   "metadata": {},
   "outputs": [],
   "source": [
    "oo_msc_5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2c62802-9de7-416f-b162-e916c9a216c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "annotations <- GetGRangesFromEnsDb(ensdb = EnsDb.Mmusculus.v79)\n",
    "seqlevelsStyle(annotations) <- \"UCSC\"\n",
    "genome(annotations) <- \"mm10\"\n",
    "Annotation(oo_msc_5) = annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5fcec8d-fd1f-4e45-a0df-e5841110c615",
   "metadata": {},
   "outputs": [],
   "source": [
    "oo_msc_5$seurat_clusters_adj = rna_seu_5@meta.data[Cells(oo_msc_5), 'seurat_clusters_adj']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a5574cde-0a83-461f-ae28-bcaebfa48171",
   "metadata": {},
   "outputs": [],
   "source": [
    "oo_msc_5$sampleID = sub('__.*', '', Cells(oo_msc_5))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd06bb11-f204-484c-8458-7243d159c64e",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 14, repr.plot.height = 4, repr.plot.res = 300)\n",
    "DimPlot(oo_msc_5, group.by = c('sampleID', 'seurat_clusters', 'seurat_clusters_adj'), pt.size = 0.1)\n",
    "# sample ID, scATAC-Seq clusters, scRNA-Seq clusters"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b3c58436-8ea0-4d09-9805-8f1a001da424",
   "metadata": {},
   "source": [
    "# plot sample vs cell numbers in each of the 4 clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ecd33b9c-0a00-4e26-86f0-4fd39fdfedcb",
   "metadata": {},
   "outputs": [],
   "source": [
    "plotdat = as.data.table(table(oo_msc_5$sampleID, oo_msc_5$seurat_clusters_adj))\n",
    "setnames(plotdat, c('sampleID', 'cluster', 'N'))\n",
    "plotdat[, cluster := factor(cluster)]\n",
    "head(plotdat)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb9876b5-fc8d-4d12-8e99-649987e9c018",
   "metadata": {},
   "outputs": [],
   "source": [
    "plotdat[N < 10, N := 0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ea9300e-cc03-4be3-a798-342435240498",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.res = 300, repr.plot.height = 4, repr.plot.width = 6)\n",
    "ggplot(plotdat, aes(x = cluster, y = N,  fill = sampleID)) + \n",
    "geom_bar(stat = 'identity', position = 'dodge') + theme_pubr() +\n",
    "ylab('Number of cells')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e5c34f1-9c8b-41a7-b481-547f4c32ce8b",
   "metadata": {},
   "source": [
    "# call peaks and plot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "252882e4-31a6-4d88-a6bd-df2a84df7632",
   "metadata": {},
   "source": [
    "## can skip this subsetion, need load oo_msc_5 in the next subsetion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00d9daac-163d-46d1-8495-f34930516807",
   "metadata": {},
   "outputs": [],
   "source": [
    "# saveRDS(oo_msc_5_peaks, file = 'data/oo_msc_5_peaks.rds')\n",
    "# oo_msc_5_peaks = readRDS(file = 'data/oo_msc_5_peaks.rds')\n",
    "oo_msc_5_peaks = CallPeaks(oo_msc_5, group.by = 'seurat_clusters_adj', macs2.path = '~/conda3/envs/mar25/bin/macs3')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65d48109-f3c2-4f59-8964-c19f52efba26",
   "metadata": {},
   "outputs": [],
   "source": [
    "oo_msc_5_peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c79444ed-a258-45c5-a274-06289da08085",
   "metadata": {},
   "outputs": [],
   "source": [
    "# organize clusters for peak\n",
    "oo_msc_5_peaks_dt = as.data.table(oo_msc_5_peaks)\n",
    "oo_msc_5_peaks_dt[, peak_called_in := paste(sort(unlist(strsplit(.SD$peak_called_in, split = ','))), collapse=','), by = 1:nrow(oo_msc_5_peaks_dt)]\n",
    "head(oo_msc_5_peaks_dt)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a3235164-5c2d-41ce-b5a8-96c3face4f3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "oo_msc_5_peaks_dt[seqnames == 'chr3' & start > 90600391 & end < 90615832, 1:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0b02d3b-6273-42ab-bc91-4f7a03d9e39f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# combine some of the clusters since these are very close WT samples\n",
    "oo_msc_5$msc_4clusters = 'cl_other'\n",
    "oo_msc_5$msc_4clusters[oo_msc_5$seurat_clusters_adj %in% c('5')]  = 'cl_5'\n",
    "oo_msc_5$msc_4clusters[oo_msc_5$seurat_clusters_adj %in% c('6')]  = 'cl_6'\n",
    "oo_msc_5$msc_4clusters[oo_msc_5$seurat_clusters_adj %in% c('7')]  = 'cl_7'\n",
    "table(oo_msc_5$msc_4clusters)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65e65acf-b8fe-4fe1-8dea-29f6d2f3ab82",
   "metadata": {},
   "outputs": [],
   "source": [
    "SplitFragments( oo_msc_5, assay = 'ATAC', group.by = 'seurat_clusters_adj',\n",
    "  outdir = getwd(), file.suffix = \"oo_msc_5_splitfragment\", outdir = 'data/', \n",
    "  append = TRUE, buffer_length = 256L, verbose = TRUE )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9c44719-4578-41b5-bd37-9537d8cba09d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Re-call peaks using MACS3 based on cluster information  \n",
    "# This is a time-consuming step  \n",
    "\n",
    "# Save the results for future use  \n",
    "# saveRDS(oo_msc_5_peaks_4clusters, file = 'data/oo_msc_5_peaks_4clusters.rds')  \n",
    "\n",
    "# Load previously saved results if needed  \n",
    "# oo_msc_5_peaks_4clusters <- readRDS(file = 'data/oo_msc_5_peaks_4clusters.rds')  \n",
    "\n",
    "# Perform peak calling with MACS3, grouping by 'msc_4clusters'  \n",
    "oo_msc_5_peaks_4clusters <- CallPeaks(  \n",
    "  object = oo_msc_5,  \n",
    "  macs2.path = '~/conda3/envs/mar25/bin/macs3',  \n",
    "  group.by = 'msc_4clusters'  \n",
    ")  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4141ca05-90b3-42fa-ad37-02f439d67ac6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# saveRDS(oo_msc_5, file = 'data/oo_msc_5_w_info.rds')\n",
    "# saveRDS(oo_msc_5_peaks_dt, file = 'data/oo_msc_5_peaks_dt.rds')\n",
    "# saveRDS(oo_msc_5_peaks, file = 'data/oo_msc_5_peaks.rds')\n",
    "# saveRDS(oo_msc_5_peaks_4clusters, file = 'data/oo_msc_5_peaks_4clusters.rds')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "013eb900-49fb-49da-8b5d-847ca6c5ed80",
   "metadata": {},
   "source": [
    "## explore the peaks (run the comment cell below if needed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d43bcbf-f98f-4823-a8eb-2b60d323f37e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# run these step if skipped the previous subsection\n",
    "# oo_msc_5 = readRDS(file = 'data/oo_msc_5_w_info.rds')\n",
    "# oo_msc_5_peaks_dt = readRDS(file = 'data/oo_msc_5_peaks_dt.rds')\n",
    "# oo_msc_5_peaks = readRDS(file = 'data/oo_msc_5_peaks.rds')\n",
    "# oo_msc_5_peaks_4clusters = readRDS(file = 'data/oo_msc_5_peaks_4clusters.rds')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "43cfc123-4070-4160-967c-847eaa7e1fd3",
   "metadata": {},
   "outputs": [],
   "source": [
    "oo_msc_5@assays$ATAC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4e900ac-b1bd-406e-a752-b886c4fbfbe3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# checking functions of this peak in other assays\n",
    "# chr3 90604391-90604832 in mm10\n",
    "subsetByOverlaps(oo_msc_5_peaks, GRanges(seqnames = 'chr3', ranges = IRanges(start = 90600882, end = 90614414)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51d32788-48cf-4bba-9b5c-0544e9336ed1",
   "metadata": {},
   "outputs": [],
   "source": [
    "oo_msc_5_peaks[13302, ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "27bf6706-ccac-4b8c-adea-70fbaa2f58ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 6, repr.plot.height = 3, repr.plot.res = 200)\n",
    "CoveragePlot(\n",
    "  object = oo_msc_5,\n",
    "  group.by = 'sampleID',\n",
    "  region = \"chr3-90600882-90614414\"\n",
    ")\n",
    "# S100a6 chr3:90,612,882-90,614,414\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3037bb6a-ac35-4c58-94d1-e2f7399b79cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 6, repr.plot.height = 4, repr.plot.res = 200)\n",
    "CoveragePlot(\n",
    "  object = oo_msc_5,\n",
    "  group.by = 'seurat_clusters',\n",
    "  region = \"chr3-90604391-90604832\"\n",
    ")\n",
    "# S100a6 chr3:90,612,882-90,614,414\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05926941-8c8c-4973-8437-ae4e045f2393",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 6, repr.plot.height = 4, repr.plot.res = 200)\n",
    "CoveragePlot(\n",
    "  object = oo_msc_5,\n",
    "  group.by = 'seurat_clusters_adj',\n",
    "  region = \"chr3-90604391-90604832\"\n",
    ")\n",
    "# S100a6 chr3:90,612,882-90,614,414\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc984e17-408d-484e-b906-947e58f47089",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 6, repr.plot.height = 4, repr.plot.res = 200)\n",
    "CoveragePlot(\n",
    "  object = oo_msc_5,\n",
    "  group.by = 'seurat_clusters_adj',\n",
    "  region = \"chr3-90600882-90614414\"\n",
    ")\n",
    "# S100a6 chr3:90,612,882-90,614,414\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ad10e4e-7aff-41ea-b727-133fa070c5f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 6, repr.plot.height = 3, repr.plot.res = 200)\n",
    "CoveragePlot(\n",
    "  object = oo_msc_5,\n",
    "  group.by = 'seurat_clusters_adj',\n",
    "  region = \"chr1-4767454-4769671\"\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a81462e5-e4de-4fe6-8267-2f71c259bbf3",
   "metadata": {},
   "outputs": [],
   "source": [
    "oo_msc_5@assays$RNA2 = rna_seu_5@assays$RNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d85ef7e7-bf53-453a-8127-08bd5ebbef11",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 6, repr.plot.height = 3, repr.plot.res = 200)\n",
    "CoveragePlot(\n",
    "    object = oo_msc_5,\n",
    "    #features = c('S100a1', 'S100a13', 'S100a16', 'S100a4', 'S100a6', 'S100a7', 'S100a8', 'S100a9'), \n",
    "    expression.assay = \"RNA2\",\n",
    "    expression.slot = \"data\",\n",
    "    group.by = 'seurat_clusters_adj',\n",
    "    region = \"chr3-90600882-90614414\"\n",
    ")\n",
    "# S100a6 chr3:90,612,882-90,614,414\n",
    "# features = c('S100a1', 'S100a13', 'S100a14', 'S100a16', 'S100a3', 'S100a4', 'S100a5', 'S100a6', 'S100a7', 'S100a8', 'S100a9'),"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf4c3132-fb52-488c-95e1-2b1821fa57db",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 12, repr.plot.height = 6, repr.plot.res = 200)\n",
    "CoveragePlot(\n",
    "  object = oo_msc_5,\n",
    "  group.by = 'seurat_clusters_adj',\n",
    "  region = \"chr3-90500000-90700000\"\n",
    ")\n",
    "# S100a6: chr3:90,612,882-90,614,414\n",
    "# S100a8: chr3:90,668,978-90,670,035\n",
    "# S100a9: chr3:90,692,632-90,695,721\n",
    "\n",
    "# S100a6:  chr3:90,612,882-90,614,414\n",
    "# S100a8: chr3:90,668,978-90,670,035\n",
    "# S100a9:chr3:90,692,632-90,695,721\n",
    "\n",
    "# chr3-90500000-90700000 \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b1673f8-5f52-4473-bd0d-123ad95a499e",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 9, repr.plot.height = 4, repr.plot.res = 200)\n",
    "CoveragePlot(\n",
    "  object = oo_msc_5,\n",
    "  group.by = 'seurat_clusters_adj',\n",
    "  region = \"chr3-90500000-90700000\"\n",
    ")\n",
    "# S100a6: chr3:90,612,882-90,614,414\n",
    "# S100a8: chr3:90,668,978-90,670,035\n",
    "# S100a9: chr3:90,692,632-90,695,721\n",
    "\n",
    "# S100a6:  chr3:90,612,882-90,614,414\n",
    "# S100a8: chr3:90,668,978-90,670,035\n",
    "# S100a9:chr3:90,692,632-90,695,721\n",
    "\n",
    "# chr3-90500000-90700000"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4732207f-584f-4bd6-a6f0-972f0496dd01",
   "metadata": {},
   "source": [
    "# peak annotate & violin plot for gene expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce5bbf59-c194-43e0-854f-34bbc98b39d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(eulerr)\n",
    "library(TxDb.Mmusculus.UCSC.mm10.knownGene)\n",
    "library(ChIPseeker)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "817c4aaf-958f-4662-944a-05a701d385f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "head(oo_msc_5_peaks)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a3e7291f-cf5a-45ad-a059-e6918b1beb58",
   "metadata": {},
   "outputs": [],
   "source": [
    "ann = annotatePeak(oo_msc_5_peaks, TxDb = TxDb.Mmusculus.UCSC.mm10.knownGene, annoDb = 'org.Mm.eg.db')\n",
    "ann_kl4 = annotatePeak(oo_msc_5_peaks_4clusters, TxDb = TxDb.Mmusculus.UCSC.mm10.knownGene, annoDb = 'org.Mm.eg.db')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3bdd8928-293f-436c-bf2c-1b8069cdaaf2",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(width = 179)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e12df05-c025-467d-8528-bfc73320e106",
   "metadata": {},
   "outputs": [],
   "source": [
    "ann"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6097f97-aef1-4567-a852-5b5bba6372f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "plotAnnoBar(ann)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb668853-21fe-41ef-8c78-e5a6e7597b30",
   "metadata": {},
   "outputs": [],
   "source": [
    "plotAnnoBar(ann_kl4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ee8ad94-77e7-45d2-8f48-0a02275441c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 12, repr.plot.height = 8, repr.plot.res = 300)\n",
    "plotAnnoPie(ann)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35c45ef1-1725-4397-a5ca-d681ef9b8f85",
   "metadata": {},
   "outputs": [],
   "source": [
    "plotDistToTSS(ann)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2bff1ef-9691-47be-84a5-28b7a182e642",
   "metadata": {},
   "outputs": [],
   "source": [
    "# display the gene expression difference of these genes with chromosome status changes\n",
    "peak_dt = as.data.table(ann@anno)\n",
    "peak_dt[, peak_called_in := paste(sort(unlist(strsplit(peak_called_in, ','))), collapse = ','), by = 1:nrow(peak_dt)]\n",
    "plotdat = peak_dt[peak_called_in %in% c('6') & abs(distanceToTSS) < 5000, .(seqnames, start, end, peak_called_in, annotation, SYMBOL, GENENAME)]\n",
    "feat_sel = intersect(plotdat$SYMBOL, rna_seu_5_markers[cluster == '6', gene])\n",
    "feat_sel\n",
    "options(repr.plot.width = 12, repr.plot.height = 6, repr.plot.res = 300)\n",
    "VlnPlot(object = rna_seu_5, features = feat_sel, ncol = 4, group.by = 'seurat_clusters_adj')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12df6fa2-99c1-42ef-992e-8cbc6132d8a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "peak_dt_kl4 = as.data.table(ann_kl4@anno)\n",
    "peak_dt_kl4[, peak_called_in := paste(sort(unlist(strsplit(peak_called_in, ','))), collapse = ','), by = 1:nrow(peak_dt_kl4)]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ecb14a1-c047-486e-87b2-26310be3f680",
   "metadata": {},
   "outputs": [],
   "source": [
    "peak_dt_kl4 = as.data.table(ann_kl4@anno)\n",
    "peak_dt_kl4[, peak_called_in := paste(sort(unlist(strsplit(peak_called_in, ','))), collapse = ','), by = 1:nrow(peak_dt_kl4)]\n",
    "plotdat = peak_dt_kl4[peak_called_in %in% c('cl_6') & abs(distanceToTSS) < 5000, .(seqnames, start, end, peak_called_in, annotation, SYMBOL, GENENAME)]\n",
    "feat_sel = intersect(plotdat$SYMBOL, rna_seu_5_markers[cluster == '6', gene])\n",
    "feat_sel\n",
    "options(repr.plot.width = 12, repr.plot.height = 6, repr.plot.res = 300)\n",
    "VlnPlot(object = rna_seu_5, features = feat_sel, ncol = 4, group.by = 'seurat_clusters_adj')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "97a12c23-1b44-40d3-8c65-b3cd9dcb4b79",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 9, repr.plot.height = 4, repr.plot.res = 200)\n",
    "CoveragePlot(\n",
    "    object = oo_msc_5,\n",
    "    group.by = 'seurat_clusters_adj',\n",
    "    region = \"chr1-171019764-171019963\",\n",
    "    extend.upstream   = 5000,\n",
    "    extend.downstream = 5000\n",
    "\n",
    ")\n",
    "#chr1\t78195219\t78195418\t6\tPromoter (1-2kb)\tPax3\tpaired box 3\n",
    "#chr1\t80710833\t80711059\t6\tPromoter (1-2kb)\tDock10\tdedicator of cytokinesis 10\n",
    "#chr1\t171019764\t171019963\t6\tPromoter (<=1kb)\tFcgr4        \t\n",
    "# lrp1b"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "43b8b5df-fee5-4832-a34c-a5e3d8c2e163",
   "metadata": {},
   "source": [
    "# overlapped of peaks from these ATACSeq cell clusters\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cd89a299-0b43-4a93-ae25-991d330d7087",
   "metadata": {},
   "outputs": [],
   "source": [
    "peak_dt = as.data.table(ann@anno)\n",
    "peak_dt[grep('1', peak_called_in), cl1 := 'Yes']\n",
    "peak_dt[grep('2', peak_called_in), cl2 := 'Yes']\n",
    "peak_dt[grep('3', peak_called_in), cl3 := 'Yes']\n",
    "peak_dt[grep('4', peak_called_in), cl4 := 'Yes']\n",
    "\n",
    "peak_dt[, icl1 := 0] \n",
    "peak_dt[, icl2 := 0] \n",
    "peak_dt[, icl3 := 0] \n",
    "peak_dt[, icl4 := 0] \n",
    "\n",
    "peak_dt[grep('1', peak_called_in), icl1 := 1] \n",
    "peak_dt[grep('2', peak_called_in), icl2 := 1]\n",
    "peak_dt[grep('3', peak_called_in), icl3 := 1]\n",
    "peak_dt[grep('4', peak_called_in), icl4 := 1]\n",
    "head( peak_dt )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb5becc8-1451-4af0-bae1-0f1b98091a94",
   "metadata": {},
   "outputs": [],
   "source": [
    "peak_dt[, nn := icl1 + icl2 + icl3 + icl4 ]\n",
    "peak_dt[, tag := paste0(seqnames, '_', start, '_', end, '_', width)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "90e45a87-71a7-4689-8700-e47f5ed493fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "s4 <- list(\n",
    "    cl1 = peak_dt[icl1 == 1, tag],\n",
    "    cl2 = peak_dt[icl2 == 1, tag],\n",
    "    cl3 = peak_dt[icl3 == 1, tag],\n",
    "    cl4 = peak_dt[icl4 == 1, tag])\n",
    "plot(venn(s4))\n",
    "plot(euler(s4, shape = \"ellipse\"), quantities = TRUE)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0223583d-4f50-4696-8e5b-b2e7549f51e2",
   "metadata": {},
   "source": [
    "# gene expression of interested genes\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3be994c3-8be5-448f-a494-6cd74c9bb94a",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 12, repr.plot.height = 6, repr.plot.res = 300)\n",
    "#VlnPlot(object = rna_seu_5, features = c('S100a1', 'S100a13', 'S100a14', 'S100a16', 'S100a3', 'S100a4', 'S100a5', 'S100a6', 'S100a7', 'S100a8', 'S100a9'), ncol = 5)\n",
    "VlnPlot(object = rna_seu_5, features = c('S100a1', 'S100a13', 'S100a16', 'S100a4', 'S100a6', 'S100a7', 'S100a8', 'S100a9'), ncol = 4, group.by = 'seurat_clusters_adj')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8c955ac-5faf-4f07-9ca0-b9ce45e8bc86",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 12, repr.plot.height = 6, repr.plot.res = 300)\n",
    "#VlnPlot(object = rna_seu_5, features = c('S100a1', 'S100a13', 'S100a14', 'S100a16', 'S100a3', 'S100a4', 'S100a5', 'S100a6', 'S100a7', 'S100a8', 'S100a9'), ncol = 5)\n",
    "VlnPlot(object = rna_seu_5, features = c('Smad1',  'Smad2',  'Smad3',  'Smad4',  'Smad5',  'Smad6', \n",
    "                                         'Smad7',  'Smad8', 'TGFBR1', 'TGFBR2'), ncol = 4, group.by = 'seurat_clusters_adj')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "58da5613-cd80-4e34-be89-f0626e782843",
   "metadata": {},
   "source": [
    "# FindAllMarkers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "780bc617-d6db-4886-a85d-da6d5b7e56e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_seu_5$seurat_clusters_adj2 = rna_seu_5$seurat_clusters_adj\n",
    "rna_seu_5$seurat_clusters_adj2[rna_seu_5$seurat_clusters_adj == 5 & rna_seu_5$umap2 < 1] = 4\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12e24c08-167e-4e39-83e7-63f1a0835604",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 9, repr.plot.height = 4, repr.plot.res = 200)\n",
    "DimPlot(rna_seu_5, group.by = \"seurat_clusters_adj\", label = TRUE) + NoLegend() + ggtitle(\"RNA\")  |\n",
    "DimPlot(rna_seu_5, group.by = \"seurat_clusters_adj2\", label = TRUE) + NoLegend() + ggtitle(\"RNA\") \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "91651a2b-232a-4f6e-b102-beebd0a04e29",
   "metadata": {},
   "outputs": [],
   "source": [
    "table(rna_seu_5$seurat_clusters_adj) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d7fbc5e-33b5-4bd6-9870-b97587bc7959",
   "metadata": {},
   "outputs": [],
   "source": [
    "Idents(rna_seu_5) = rna_seu_5$seurat_clusters_adj2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e73f940e-1416-4044-8751-2c9ee1809138",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# saved var\n",
    "rna_seu_5_markers = FindAllMarkers(rna_seu_5, only.pos = T)\n",
    "rna_seu_5_markers = as.data.table(rna_seu_5_markers, keep.rownames = T)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa87968a-c6c4-4f7c-be6d-573c2ea2047e",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_seu_5_markers[gene == 'S100a6', ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c94dd638-c1b6-4436-8b15-d136569c68a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_seu_5_markers %>% group_by(cluster) %>% head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed813212-ffd7-4150-b103-66c0091cb534",
   "metadata": {},
   "outputs": [],
   "source": [
    "feat_sel = unique(c(\n",
    "    rna_seu_5_markers[cluster %in% c('1') ][order(p_val_adj), gene][1:5],\n",
    "    rna_seu_5_markers[cluster %in% c('2') ][order(p_val_adj), gene][1:5],\n",
    "    rna_seu_5_markers[cluster %in% c('3') ][order(p_val_adj), gene][1:5],\n",
    "    rna_seu_5_markers[cluster %in% c('4') ][order(p_val_adj), gene][1:5],\n",
    "    rna_seu_5_markers[cluster %in% c('5') ][order(p_val_adj), gene][1:5],\n",
    "    rna_seu_5_markers[cluster %in% c('6') ][order(p_val_adj), gene][1:5],\n",
    "    'S100a6',\n",
    "    rna_seu_5_markers[cluster %in% c('7') ][order(p_val_adj), gene][1:50]) )\n",
    "head(feat_sel)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef2adee2-fbb3-44d5-815f-a70484a52023",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_seu_5_copy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21a990e7-48d4-47e5-b049-5e2e4037ce69",
   "metadata": {},
   "outputs": [],
   "source": [
    "dim(rna_seu_5_copy@assays$RNA$scale.data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18e379a8-a49a-4591-80d1-bc8f7e3c9042",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 10, repr.plot.height = 8, repr.plot.res = 200)\n",
    "rna_seu_5_copy = ScaleData(rna_seu_5, features = feat_sel , assay = 'RNA') \n",
    "ret = DoHeatmap(rna_seu_5_copy, features = feat_sel, group.by = 'seurat_clusters_adj', assay = 'RNA', slot = 'scale.data') + \n",
    "scale_fill_gradientn(colors = c(\"blue\", \"white\", \"red\"))\n",
    "ret\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7dfb304b-152c-4670-a9b7-3eb8c37ca8ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "head(rna_seu_5_markers)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d9709850-384f-42e2-941c-8dc90d3af0a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_seu_5_copy = copy(rna_seu_5)\n",
    "Idents(rna_seu_5_copy) = rna_seu_5$kl\n",
    "rna_seu_5_markers_kl34 = FindAllMarkers(rna_seu_5_copy, only.pos = T)\n",
    "rna_seu_5_markers_kl34 = as.data.table(rna_seu_5_markers_kl34, keep.rownames = T)\n",
    "head(rna_seu_5_markers_kl34)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5239da17-0114-441c-bac9-5617fb647f21",
   "metadata": {},
   "source": [
    "# MS5 cell bulk rna seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "66edadcd-9d1a-4b17-b411-a25597d0be11",
   "metadata": {},
   "outputs": [],
   "source": [
    "# to validate these results, a bulk ms5 (mouse MSC cell) bulk RNA-Seq was introduced\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df3df4db-cf02-43ed-a074-a2abc14e5416",
   "metadata": {},
   "outputs": [],
   "source": [
    "ms5_deg = fread('~/WenhuoHu/collab_JinWang/ms5/deg.xls')\n",
    "head(ms5_deg) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49274336-023e-48c2-a8db-f234499bf70d",
   "metadata": {},
   "outputs": [],
   "source": [
    "fwrite(rna_seu_5_markers, file = '~/WenhuoHu/collab_JinWang/rna_seu_5_markers.tsv', sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "234ab40d-ee44-44b5-8bf6-2dec1435d2bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "head(rna_seu_5_markers)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "89f8be48-0dc2-4cdb-8fc6-b3ac9137e678",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(ggVennDiagram)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0365d63-ab39-4357-9798-809abafee9b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_sel1 = unique(rna_seu_5_markers[p_val_adj < 0.05, gene]) \n",
    "gene_sel2 = unique(ms5_deg[FDR < 0.05, rn])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95ecd25d-cb2e-467e-ab7a-e0da89466d78",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 6, repr.plot.height = 4, repr.plot.res = 200)\n",
    "ggVennDiagram(list(gene_sel1, gene_sel2)) + coord_flip()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d6867542-40ee-4430-ae1a-f6bf064765e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 10, repr.plot.height = 8, repr.plot.res = 200)\n",
    "gene_sel1 = unique(rna_seu_5_markers[p_val_adj < 0.05 & abs(avg_log2FC) > 1, gene]) \n",
    "gene_sel2 = unique(ms5_deg[FDR < 0.05 & abs(logFC) > 1, rn])\n",
    "feat_sel = c(intersect(gene_sel1, gene_sel2), 'S100a6')\n",
    "rna_seu_5_copy = ScaleData(rna_seu_5, features = feat_sel , assay = 'RNA') \n",
    "ret = DoHeatmap(rna_seu_5_copy, features = feat_sel, group.by = 'seurat_clusters_adj2', assay = 'RNA', slot = 'scale.data') + \n",
    "scale_fill_gradientn(colors = c(\"blue\", \"white\", \"red\"))\n",
    "ret\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "931ea284-85a5-497c-bdd8-7931d238da07",
   "metadata": {},
   "outputs": [],
   "source": [
    "logcpm = read.table('ms5/logcpm_ms5.tsv') \n",
    "ms5_dsn = read.table('ms5/dsn.tsv') "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "476b67b7-c92e-455e-9dde-8fb8b69e6bc8",
   "metadata": {},
   "outputs": [],
   "source": [
    "feat_sel = c(intersect(gene_sel1, gene_sel2), 'S100a6')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5d1f8df6-9007-47c1-869d-1187a54b8c58",
   "metadata": {},
   "outputs": [],
   "source": [
    "plotdat = logcpm[feat_sel, ms5_dsn$sample_name] \n",
    "plotdat = scale_fun(plotdat)\n",
    "plotdat[plotdat > 2] = 2\n",
    "plotdat[plotdat < -2] = -2\n",
    "dim(plotdat)\n",
    "\n",
    "column_ha = HeatmapAnnotation(df = ms5_dsn[, 'group', drop = F], show_annotation_name = F)\n",
    "\n",
    "options(repr.plot.width = 4, repr.plot.height = 5, repr.plot.res = 300)\n",
    "Heatmap(plotdat, show_column_names = F, show_row_names = T, show_column_dend = F, cluster_columns  = F, \n",
    "        cluster_rows = T,   row_names_gp = grid::gpar(fontsize = 4),   ##row_split = factor(dsn_row$target),\n",
    "        show_row_dend = F, top_annotation = column_ha, column_split = ms5_dsn$group) \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35c91747-4ead-407f-8ef7-7e1a04b78b51",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta_gene = read_xlsx('~/Mammalian_Metabolic_Final.xlsx')\n",
    "setnames(meta_gene, 1, 'gene')\n",
    "head(meta_gene) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7ad318d-a153-428e-9802-dc6d51c80785",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_sel1 = unique(rna_seu_5_markers[p_val_adj < 0.05 & abs(avg_log2FC) > 0 & cluster %in% c(5,6,7), gene]) \n",
    "gene_sel2 = unique(ms5_deg[FDR < 0.05 & abs(logFC) > 0, rn])\n",
    "intersect(meta_gene$gene, gene_sel1)\n",
    "intersect(meta_gene$gene, gene_sel2)\n",
    "intersect(intersect(meta_gene$gene, gene_sel2), intersect(meta_gene$gene, gene_sel1)) \n",
    "feat_sel = intersect(intersect(meta_gene$gene, gene_sel2), intersect(meta_gene$gene, gene_sel1)) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58c0bffa-7bc0-4237-8cf9-0f35d1c8ad92",
   "metadata": {},
   "outputs": [],
   "source": [
    "plotdat = logcpm[feat_sel, ms5_dsn$sample_name] \n",
    "plotdat = scale_fun(plotdat)\n",
    "plotdat[plotdat > 2] = 2\n",
    "plotdat[plotdat < -2] = -2\n",
    "dim(plotdat)\n",
    "\n",
    "column_ha = HeatmapAnnotation(df = ms5_dsn[, 'group', drop = F], show_annotation_name = F)\n",
    "\n",
    "options(repr.plot.width = 4, repr.plot.height = 3, repr.plot.res = 300)\n",
    "Heatmap(plotdat, show_column_names = F, show_row_names = T, show_column_dend = F, cluster_columns  = F, \n",
    "        cluster_rows = T,   row_names_gp = grid::gpar(fontsize = 6),   ##row_split = factor(dsn_row$target),\n",
    "        show_row_dend = F, top_annotation = column_ha, column_split = ms5_dsn$group) \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24c7c5f4-dade-4aa6-9664-5e4757e40bfc",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 7, repr.plot.height = 5, repr.plot.res = 200)\n",
    "rna_seu_5_copy = ScaleData(rna_seu_5, features = feat_sel , assay = 'RNA') \n",
    "ret = DoHeatmap(rna_seu_5_copy, features = feat_sel, group.by = 'seurat_clusters_adj2', assay = 'RNA', slot = 'scale.data') + \n",
    "scale_fill_gradientn(colors = c(\"blue\", \"white\", \"red\"))\n",
    "ret\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9544f187-fc9c-4e64-850e-7e97feb3ee47",
   "metadata": {},
   "outputs": [],
   "source": [
    "ms5_deg[rn %in% c('Aoc3'), ]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "949c12c1-bfdc-41a4-b199-080c15fa2ae1",
   "metadata": {},
   "source": [
    "# cell cycle scoring"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "173ab498-b7e6-4714-b887-aec256acba83",
   "metadata": {},
   "outputs": [],
   "source": [
    "# conclustion: no obvious cell cycle difference of these MSC clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6fa61d18-1272-421f-b0ef-f02789e147f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "geneId = as.data.table(read.csv('~/WenhuoHu/repos/niche/data/human_to_mouse.csv'))\n",
    "setkey(geneId, 'Gene.name')\n",
    "head(geneId)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cddcc820-6fed-412e-a07c-77164e000648",
   "metadata": {},
   "outputs": [],
   "source": [
    "m.s.genes = geneId[cc.genes$s.genes, Mouse.gene.name]\n",
    "m.g2m.genes = geneId[cc.genes$g2m.genes, Mouse.gene.name]\n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4bcee33c-8a21-43bc-8f34-d920be1b8777",
   "metadata": {},
   "outputs": [],
   "source": [
    "m.s.genes = m.s.genes[!is.na(m.s.genes) ]\n",
    "m.g2m.genes = m.g2m.genes[!is.na(m.g2m.genes) ]\n",
    "m.s.genes = m.s.genes[m.s.genes != '' ]\n",
    "m.g2m.genes = m.g2m.genes[m.g2m.genes != '' ]\n",
    "m.s.genes\n",
    "m.g2m.genes\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7fe105e6-47ea-4881-8f26-b6fa194ad0ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "DefaultAssay(rna_seu_5) = 'RNA'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e693d5d3-9507-4742-9a43-9aac2eafbb2c",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_seu_5 <- CellCycleScoring(rna_seu_5, s.features = m.s.genes, g2m.features = m.g2m.genes, set.ident = TRUE, group.by = 'seurat_clusters_adj')\n",
    "\n",
    "# view cell cycle scores and phase assignments\n",
    "head(rna_seu_5@meta.data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "90ab3b6d-e1cf-4164-885a-ae32319df4bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "DimPlot(rna_seu_5, group.by = 'Phase')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7721d48-662c-4f91-9df5-98f6b0dee70e",
   "metadata": {},
   "outputs": [],
   "source": [
    "plotdat = as.data.table(rna_seu_5@meta.data)\n",
    "plotdat = as.data.table(table(plotdat$Phase, plotdat$seurat_clusters_adj)) \n",
    "plotdat[, cluster_nn := sum(.SD$N), by = 'V2']\n",
    "plotdat[, percentage := N / cluster_nn ]\n",
    "setnames(plotdat, 'V1', 'Phase')\n",
    "head(plotdat)\n",
    "ggbarplot(plotdat, x = 'V2', y = 'percentage', fill = 'Phase') + xlab('Clusters') + ylab('Ratio in each cluster')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e722831d-9f20-4a8a-bd1d-21e6e6ca4365",
   "metadata": {},
   "source": [
    "# genes related with osteoblast cell functions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9877ebde-17f6-449e-a5e5-4da9199b9fb3",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## David Scaden's results (deprecated)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8df2376a-4d53-476a-89a4-c954dbfc5def",
   "metadata": {},
   "outputs": [],
   "source": [
    "# raw data of GSE128432\n",
    "# /research_jude/rgs01_jude/groups/jxugrp/home/common/Data_Public/scRNA_seq/GSE128432/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "543a93a0-832d-4f31-9fcb-39e9653d5ed9",
   "metadata": {},
   "outputs": [],
   "source": [
    "getwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2fdaf541-d7e0-457a-8965-91bc7ca48251",
   "metadata": {},
   "outputs": [],
   "source": [
    "setwd('/research_jude/rgs01_jude/groups/jxugrp/home/common/Data_Public/scRNA_seq/GSE128432/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31c52af2-d513-4f9f-881b-38a0880a1685",
   "metadata": {},
   "outputs": [],
   "source": [
    "# https://github.com/pblaney/scXploringTheNiche/blob/main/scRnaTaxonomyOfBmStromaHomeostasis.Rmd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f549341a-176a-4cce-bb89-7f66484fbf58",
   "metadata": {},
   "outputs": [],
   "source": [
    "std1 <- Read10X(data.dir = \"std1/\")\n",
    "std2 <- Read10X(data.dir = \"std2/\")\n",
    "std3 <- Read10X(data.dir = \"std3/\")\n",
    "std4 <- Read10X(data.dir = \"std4/\")\n",
    "std5 <- Read10X(data.dir = \"std5/\")\n",
    "std6 <- Read10X(data.dir = \"std6/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5cf53c19-f238-4dcc-acae-68432aea3aa8",
   "metadata": {},
   "outputs": [],
   "source": [
    "std1 <- CreateSeuratObject(counts = std1, project = \"std1\", assay = \"RNA\")\n",
    "std2 <- CreateSeuratObject(counts = std2, project = \"std2\", assay = \"RNA\")\n",
    "std3 <- CreateSeuratObject(counts = std3, project = \"std3\", assay = \"RNA\")\n",
    "std4 <- CreateSeuratObject(counts = std4, project = \"std4\", assay = \"RNA\")\n",
    "std5 <- CreateSeuratObject(counts = std5, project = \"std5\", assay = \"RNA\")\n",
    "std6 <- CreateSeuratObject(counts = std6, project = \"std6\", assay = \"RNA\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7e29eac-d905-4890-a29d-6d14c48f5221",
   "metadata": {},
   "outputs": [],
   "source": [
    "std_oo = list(std1, std2, std3, std4, std5, std6)\n",
    "std_oo = lapply(std_oo, function(oo){\n",
    "    oo[[\"Percent_MT\"]] <- PercentageFeatureSet(object = oo, pattern = \"^mt-\")\n",
    "    oo\n",
    "    #subset(x = oo, subset = nFeature_RNA < 5000 & nFeature_RNA > 500 & nCount_RNA < 32500 & Percent_MT < 12)\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "060f8ab7-de29-4cda-8196-2f2fb11ce690",
   "metadata": {},
   "outputs": [],
   "source": [
    "std_oo[[1]] = subset(x = std_oo[[1]], subset = nFeature_RNA < 5000 & nFeature_RNA > 500 & nCount_RNA < 32500 & Percent_MT < 12)\n",
    "std_oo[[2]] = subset(x = std_oo[[1]], subset = nFeature_RNA < 3250 & nFeature_RNA > 500 & nCount_RNA < 15000 & Percent_MT < 10)\n",
    "std_oo[[3]] = subset(x = std_oo[[1]], subset = nFeature_RNA < 3000 & nFeature_RNA > 500 & nCount_RNA < 12500 & Percent_MT < 10)\n",
    "std_oo[[4]] = subset(x = std_oo[[1]], subset = nFeature_RNA < 4250 & nFeature_RNA > 500 & nCount_RNA < 20000 & Percent_MT < 10)\n",
    "std_oo[[5]] = subset(x = std_oo[[1]], subset = nFeature_RNA < 4500 & nFeature_RNA > 500 & nCount_RNA < 30000 & Percent_MT < 10)\n",
    "std_oo[[6]] = subset(x = std_oo[[1]], subset = nFeature_RNA < 4000 & nFeature_RNA > 500 & nCount_RNA < 18000 & Percent_MT < 10)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36536e90-a9c2-4237-bb6b-3506dcf80a26",
   "metadata": {},
   "outputs": [],
   "source": [
    "std_oo <- lapply(X = std_oo, FUN =  function(x) {\n",
    "    x <- NormalizeData(object = x,\n",
    "                     normalization.method = \"LogNormalize\",\n",
    "                     scale.factor = 10000,\n",
    "                     verbose = FALSE)\n",
    "\n",
    "    x <- FindVariableFeatures(object = x,\n",
    "                            selection.method = \"vst\",\n",
    "                            nfeatures = 5000,\n",
    "                            verbose = FALSE)\n",
    "    x\n",
    "})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9278291d-379e-4bc0-80a2-25ddaffc0c73",
   "metadata": {},
   "outputs": [],
   "source": [
    "stdIntegrationFeatures <- SelectIntegrationFeatures(object.list = std_oo, nfeatures = 5000)\n",
    "\n",
    "std_oo <- lapply(X = std_oo, FUN = function(x) {\n",
    "    x <- ScaleData(object = x, features = stdIntegrationFeatures, verbose = FALSE)\n",
    "\n",
    "    x <- RunPCA(object = x, features = stdIntegrationFeatures, verbose = FALSE)\n",
    "    x\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59414f88-7418-4e9b-af7f-38b46e0ff291",
   "metadata": {},
   "outputs": [],
   "source": [
    "stdIntegrationAnchors <- FindIntegrationAnchors(object.list = std_oo, reduction = \"rpca\", anchor.features = stdIntegrationFeatures)\n",
    "stdIntegrated <- IntegrateData(anchorset = stdIntegrationAnchors)\n",
    "stdIntegrated <- ScaleData(object = stdIntegrated, verbose = FALSE)\n",
    "\n",
    "# Space clean up\n",
    "rm(std1, std2, std3, std4, std5, std6, stdList)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "384bced1-4852-42e9-8a0d-2e1ad2f4bb4f",
   "metadata": {},
   "outputs": [],
   "source": [
    "stdIntegrated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6474e03d-7c41-446a-9662-ce1d6a71ded2",
   "metadata": {},
   "outputs": [],
   "source": [
    "allGenes <- rownames(stdIntegrated@assays$RNA$counts)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4418c0ed-7361-480e-ad2d-1c98f96144db",
   "metadata": {},
   "outputs": [],
   "source": [
    "length(allGenes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "684dd4c6-b1e4-4d72-ad1a-c75100504ec8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# First filter out all mitochondrial genes by finding all included in original gene\n",
    "# list, then checking the variable feature list to see if any of those genes were included.\n",
    "# If so, remove them from the variable feature set.\n",
    "mitochondrialGenes <- grep(pattern = \"mt-\",\n",
    "                           x = allGenes,\n",
    "                           ignore.case = TRUE,\n",
    "                           value = TRUE\n",
    "                          )\n",
    "\n",
    "# Remove the mitochondrial genes from the variable features list\n",
    "stdIntegrated[['integrated']]@var.features <- setdiff(x = stdIntegrated[['integrated']]@var.features,\n",
    "                                                      y = mitochondrialGenes\n",
    "                                                     )\n",
    "\n",
    "# To remove ribosomal protein genes, read in gene list sourced from the HUGO Gene\n",
    "# Nomenclature Committee (HGNC) and parse it to isolate gene names\n",
    "hgncRibosomalProteinGeneList <- read_delim(file = \"~/repos/scXploringTheNiche/data/hgncRibosomalProteinGeneList.txt\",\n",
    "                                           col_names = TRUE, \n",
    "                                           delim = \"\\t\"\n",
    "                                          )$`Approved symbol`\n",
    "\n",
    "# Loop through ribosomal protein gene list and determine if that gene is listed\n",
    "# in the original full gene list\n",
    "ribosomalProteinGenes <- character()\n",
    "for(i in 1:length(hgncRibosomalProteinGeneList)) {\n",
    "    geneMatch <- grep(\n",
    "        pattern = paste(\"^\", hgncRibosomalProteinGeneList[i], \"$\", sep = \"\"),\n",
    "        x = allGenes,\n",
    "        ignore.case = TRUE,\n",
    "        value = TRUE,\n",
    "        perl = TRUE\n",
    "    )\n",
    "    ribosomalProteinGenes <- c(ribosomalProteinGenes, geneMatch)\n",
    "}\n",
    "ribosomalProteinGenes <- unique(ribosomalProteinGenes)\n",
    "\n",
    "# Remove the ribosomal protein genes from the variable features list\n",
    "stdIntegrated[['integrated']]@var.features <- setdiff(x = stdIntegrated[['integrated']]@var.features,\n",
    "                                                      y = ribosomalProteinGenes\n",
    "                                                     )\n",
    "\n",
    "# To remove cell cycle genes, read in gene list sourced from KEGG/Broad Institute\n",
    "keggCellCycleGeneList <- as_vector(\n",
    "    read.csv(\"~/repos/scXploringTheNiche/data/keggGeneSet.txt\", skip = 2, col.names = 'gene')\n",
    ")\n",
    "\n",
    "# Loop through cell cycle gene list and determine if that gene is listed\n",
    "# in the original full gene list\n",
    "cellCycleGenes <- character()\n",
    "for(i in 1:length(keggCellCycleGeneList)) {\n",
    "    geneMatch <- grep(\n",
    "        pattern = paste(\"^\", keggCellCycleGeneList[i], \"$\", sep = \"\"),\n",
    "        x = allGenes,\n",
    "        ignore.case = TRUE,\n",
    "        value = TRUE,\n",
    "        perl = TRUE\n",
    "    )\n",
    "    cellCycleGenes <- c(cellCycleGenes, geneMatch)\n",
    "}\n",
    "\n",
    "# Remove the cell cycle genes from the variable features list\n",
    "stdIntegrated[['integrated']]@var.features <- setdiff(x = stdIntegrated[['integrated']]@var.features,\n",
    "                                                      y = cellCycleGenes\n",
    "                                                     )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36260867-afe5-45f3-82fa-0b9d419b2ba6",
   "metadata": {},
   "outputs": [],
   "source": [
    "stdIntegrated <- RunPCA(object = stdIntegrated, verbose = FALSE, npcs = 65)\n",
    "\n",
    "# Produce a PCA plot where each point represents a cell and its position relative to the\n",
    "# origin point (0,0) shows how much influence it has on the that particular PC. Additionally,\n",
    "# the correlation between two points can be determined by the angle between the vectors from\n",
    "# each point to the origin point. Smaller the angle, the more positive correlation.\n",
    "# Perpendicular angles mean likely no correlation.\n",
    "DimPlot(object = stdIntegrated, reduction = \"pca\")\n",
    "\n",
    "# Display a heatmap showing how genes correlate with what PC. Each line represents a cell\n",
    "# and the color represents the loading of the gene (the genes shown are most extremely variable)\n",
    "# with black being 0, yellow being positive and purple being negative.\n",
    "DimHeatmap(object = stdIntegrated, dims = 1, balanced = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "850d5b36-6df1-487a-8079-d43d637a13ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "ElbowPlot( object = stdIntegrated,  ndims = 65) + geom_abline( aes(intercept = 1.75, slope = 0, color = \"red\"), show.legend = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1922e028-c8cd-4210-bbfb-6da8f52a13ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "stdIntegrated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e1278151-1e95-49df-8e7f-041e0a340ea1",
   "metadata": {},
   "outputs": [],
   "source": [
    "names(stdIntegrated@graphs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e84a9d5f-5ddc-44a2-a58e-3a90cd690024",
   "metadata": {},
   "outputs": [],
   "source": [
    "stdIntegrated <- FindNeighbors(object = stdIntegrated, dims = 1:50)\n",
    "stdIntegrated <- FindClusters(object = stdIntegrated, resolution = 0.64, random.seed = 2019, reduction = 'pca')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "127a6408-a00f-4169-9663-2b38ee98550d",
   "metadata": {},
   "outputs": [],
   "source": [
    "stdIntegrated <- RunTSNE(object = stdIntegrated, dims = 1:50)\n",
    "\n",
    "# Plot the tSNE reduction grouped by sample origin then by cluster\n",
    "DimPlot(object = stdIntegrated, reduction = \"tsne\", group.by = 'orig.ident')\n",
    "\n",
    "DimPlot(object = stdIntegrated, reduction = \"tsne\", label = TRUE, label.size = 5,\n",
    "        cols = c(\"aquamarine2\", \"blueviolet\", \"chartreuse2\", \"darkgreen\", \"darkred\", \n",
    "                 \"darkgray\", \"chocolate1\", \"coral4\", \"blue\", \"darkmagenta\", \"deeppink2\",\n",
    "                 \"goldenrod1\", \"yellow1\", \"tan4\", \"tomato1\", \"olivedrab\", \"navajowhite2\",\n",
    "                 \"orangered\", \"palegreen\", \"peru\", \"plum1\", \"purple3\", \"red2\", \"salmon\",\n",
    "                 \"seagreen\", \"magenta\", \"midnightblue\", \"khaki\", \"greenyellow\", \"deeppink\",\n",
    "                 \"burlywood3\", \"dimgray\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "553d6f24-5c6e-469b-9e56-1dd23a27b8ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "saveRDS(object = stdIntegrated, file = \"data/stdIntegrated.rds\")\n",
    "#stdIntegrated <- readRDS(file = \"data/stdIntegrated.rds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6bcba9a5-fae0-4118-9c67-dd9393e4512b",
   "metadata": {},
   "outputs": [],
   "source": [
    "DefaultAssay(stdIntegrated) <- \"RNA\"\n",
    "FeaturePlot(object = stdIntegrated,\n",
    "            features = c(\"Cxcl12\", \"Bglap\", \"Cdh5\", \"Acan\", \"S100a4\", \"Acta2\"),\n",
    "            reduction = \"tsne\",\n",
    "            label = TRUE,\n",
    "            ncol = 3,\n",
    "            cols = c(\"lightskyblue1\", \"red3\"))\n",
    "\n",
    "# Plot the distribution of expression for all cells related to specific genes\n",
    "# with a violin plot\n",
    "VlnPlot(object = stdIntegrated,\n",
    "        features = c(\"Cxcl12\", \"Bglap\", \"Cdh5\", \"Acan\", \"S100a4\", \"Acta2\"),\n",
    "        pt.size = 0.5,\n",
    "        ncol = 2)\n",
    "DefaultAssay(stdIntegrated) <- \"integrated\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "599103a6-b54f-4210-bfb6-7d672d24e1f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Visualize the distribution of cells expressing gene markers across all clusters\n",
    "DefaultAssay(stdIntegrated) <- \"RNA\"\n",
    "FeaturePlot(object = stdIntegrated,\n",
    "            features = c(\"Cd79a\", \"Cd79b\"),\n",
    "            reduction = \"tsne\",\n",
    "            label = TRUE,\n",
    "            label.size = 4,\n",
    "            cols = c(\"lightskyblue1\", \"red3\"))\n",
    "# Appears to be 7, 9, 12, 25, 27 \n",
    "VlnPlot(object = stdIntegrated,\n",
    "        features = c(\"Cd79a\", \"Cd79b\"),\n",
    "        pt.size = 0.5, \n",
    "        idents = c(7, 9, 12, 25, 27, 1, 2))\n",
    "\n",
    "### NOTE ###\n",
    "# These methods are loose and not rooted in any statistical rule, depreceated\n",
    "# Determine cutoff values for expression levels of gene markers based on expression across all cells\n",
    "#markerCd79a <- stdIntegrated[[\"integrated\"]]@scale.data[\"Cd79a\",]\n",
    "#quantile(markerCd79a, probs = seq(0.6, 1, 0.025))\n",
    "\n",
    "#markerCd79b <- stdIntegrated[[\"integrated\"]]@scale.data[\"Cd79b\",]\n",
    "#quantile(markerCd79b, probs = seq(0.6, 1, 0.025))\n",
    "\n",
    "# Identify all the cells that satisfy cutoff value which was selected to be between 85 and 90 quantile\n",
    "#clusterFilter1 <- WhichCells(object = stdIntegrated,\n",
    "#                             expression = Cd79a > 1 & Cd79b > 1)\n",
    "\n",
    "# Visualize count of cells per cluster\n",
    "#clusterFilter1 <- table(stdIntegrated$seurat_clusters[clusterFilter1])\n",
    "\n",
    "bcellClusters <- c(7, 9, 12, 25, 27)\n",
    "DefaultAssay(stdIntegrated) <- \"integrated\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "481ef538-0146-4ac3-b9f2-6f94d9af7f71",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Visualize the distribution of cells expressing gene markers across all clusters\n",
    "DefaultAssay(stdIntegrated) <- \"RNA\"\n",
    "FeaturePlot(object = stdIntegrated,\n",
    "            features = c(\"Gypa\", \"Hbb-bt\", \"Hbb-bs\", \"Rhag\", \"Rhd\", \"Tfrc\"),\n",
    "            reduction = \"tsne\",\n",
    "            label = TRUE,\n",
    "            label.size = 4,\n",
    "            cols = c(\"lightskyblue1\", \"red3\"))\n",
    "# looks like 5, 18\n",
    "VlnPlot(object = stdIntegrated,\n",
    "        features = c(\"Gypa\", \"Hbb-bt\", \"Hbb-bs\", \"Rhag\", \"Rhd\", \"Tfrc\"),\n",
    "        pt.size = 0.5, \n",
    "        idents = c(5, 18, 0, 1, 2))\n",
    "\n",
    "# Determine cutoff values for expression levels of gene markers based on expression across all cells\n",
    "#markerGypa <- stdIntegrated[[\"integrated\"]]@scale.data[\"Gypa\",]\n",
    "#quantile(markerGypa, probs = seq(0.6, 1, 0.025))\n",
    "\n",
    "#markerHbbbt <- stdIntegrated[[\"integrated\"]]@scale.data[\"Hbb-bt\",]\n",
    "#quantile(markerHbbbt, probs = seq(0.6, 1, 0.025))\n",
    "\n",
    "#markerHbbbs <- stdIntegrated[[\"integrated\"]]@scale.data[\"Hbb-bs\",]\n",
    "#quantile(markerHbbbs, probs = seq(0.6, 1, 0.025))\n",
    "\n",
    "#markerRhag <- stdIntegrated[[\"integrated\"]]@scale.data[\"Rhag\",]\n",
    "#quantile(markerRhag, probs = seq(0.6, 1, 0.025))\n",
    "\n",
    "#markerRhd <- stdIntegrated[[\"integrated\"]]@scale.data[\"Rhd\",]\n",
    "#quantile(markerRhd, probs = seq(0.6, 1, 0.025))\n",
    "\n",
    "#markerTfrc <- stdIntegrated[[\"integrated\"]]@scale.data[\"Tfrc\",]\n",
    "#quantile(markerTfrc, probs = seq(0.6, 1, 0.025))\n",
    "\n",
    "# Identify all the cells that satisfy cutoff value which was selected based on distribution\n",
    "#clusterFilter2 <- WhichCells(\n",
    "#        object = stdIntegrated,\n",
    "#        expression = Gypa > 0.15 & \"Hbb-bt\" > 0.15 & \"Hbb-bs\" > 0.15 & Rhag > 0.15 & Rhd > 0.15 & Tfrc > 0.15)\n",
    "\n",
    "# Visualize count of cells per cluster\n",
    "#clusterFilter2 <- table(stdIntegrated$seurat_clusters[clusterFilter2])\n",
    "\n",
    "\n",
    "erythroidClusters <- c(5,18)\n",
    "DefaultAssay(stdIntegrated) <- \"integrated\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72315895-3f6c-42c0-94a4-4036d70e66a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "DefaultAssay(stdIntegrated) <- \"RNA\"\n",
    "FeaturePlot(object = stdIntegrated,\n",
    "            features = c(\"Cd52\", \"Cd177\", \"Plaur\", \"Clec4a2\"),\n",
    "            reduction = \"tsne\",\n",
    "            label = TRUE,\n",
    "            label.size = 4,\n",
    "            cols = c(\"lightskyblue1\", \"red3\"))\n",
    "# looks like 14, 16, 31\n",
    "VlnPlot(object = stdIntegrated,\n",
    "        features = c(\"Cd52\", \"Cd177\", \"Plaur\", \"Clec4a2\"),\n",
    "        pt.size = 0.5, \n",
    "        idents = c(14, 16, 31, 0, 1, 2),\n",
    "        ncol = 2)\n",
    "\n",
    "# Determine cutoff values for expression levels of gene markers based on expression across all cells\n",
    "#markerCd52 <- stdIntegrated[[\"integrated\"]]@scale.data[\"Cd52\",]\n",
    "#quantile(markerCd52, probs = seq(0.6, 1, 0.025))\n",
    "\n",
    "#markerCd177 <- stdIntegrated[[\"integrated\"]]@scale.data[\"Cd177\",]\n",
    "#quantile(markerCd177, probs = seq(0.6, 1, 0.025))\n",
    "\n",
    "#markerPlaur <- stdIntegrated[[\"integrated\"]]@scale.data[\"Plaur\",]\n",
    "#quantile(markerPlaur, probs = seq(0.6, 1, 0.025))\n",
    "\n",
    "#markerClec4a2 <- stdIntegrated[[\"integrated\"]]@scale.data[\"Clec4a2\",]\n",
    "#quantile(markerClec4a2, probs = seq(0.6, 1, 0.025))\n",
    "\n",
    "# Identify all the cells that satisfy cutoff value which was selected based on distribution\n",
    "#clusterFilter3 <- WhichCells(object = stdIntegrated,\n",
    "#                             expression = Cd52 > 0.4 & Cd177 > 0.075 & Plaur > 0.25 & Clec4a2 > 0.5)\n",
    "\n",
    "# Visualize count of cells per cluster\n",
    "#clusterFilter3 <-table(stdIntegrated$seurat_clusters[clusterFilter3])\n",
    "\n",
    "granulocyteClusters <- c(14, 16, 31)\n",
    "DefaultAssay(stdIntegrated) <- \"integrated\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7cb7176c-8add-4d36-beb3-6cfc6f8cf8d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Visualize the distribution of cells expressing gene markers across all clusters\n",
    "DefaultAssay(stdIntegrated) <- \"RNA\"\n",
    "FeaturePlot(object = stdIntegrated,\n",
    "            features = c(\"Gp9\", \"Itga2b\", \"Cd9\", \"Gp1bb\"),\n",
    "            reduction = \"tsne\",\n",
    "            label = TRUE,\n",
    "            label.size = 4,\n",
    "            cols = c(\"lightskyblue1\", \"red3\"))\n",
    "# looks like 21, 28\n",
    "VlnPlot(object = stdIntegrated,\n",
    "        features = c(\"Gp9\", \"Itga2b\", \"Cd9\", \"Gp1bb\"),\n",
    "        pt.size = 0.5,\n",
    "        idents = c(21, 28, 0, 1, 2),\n",
    "        ncol = 2)\n",
    "\n",
    "# Determine cutoff values for expression levels of gene markers based on expression across all cells\n",
    "#markerGp9 <- stdIntegrated[[\"integrated\"]]@scale.data[\"Gp9\",]\n",
    "#quantile(markerGp9, probs = seq(0.6, 1, 0.025))\n",
    "\n",
    "#markerItga2b <- stdIntegrated[[\"integrated\"]]@scale.data[\"Itga2b\",]\n",
    "#quantile(markerItga2b, probs = seq(0.6, 1, 0.025))\n",
    "\n",
    "#markerCd9 <- stdIntegrated[[\"integrated\"]]@scale.data[\"Cd9\",]\n",
    "#quantile(markerCd9, probs = seq(0.6, 1, 0.025))\n",
    "\n",
    "#markerGp1bb <- stdIntegrated[[\"integrated\"]]@scale.data[\"Gp1bb\",]\n",
    "#quantile(markerGp1bb, probs = seq(0.6, 1, 0.025))\n",
    "\n",
    "# Identify all the cells that satisfy cutoff value which was selected based on distribution\n",
    "#clusterFilter5 <- WhichCells(object = stdIntegrated,\n",
    "#                             expression = Gp9 > 0.75 & Itga2b > 0.75 & Cd9 > 1.25 & Gp1bb > 1)\n",
    "\n",
    "# Visualize count of cells per cluster\n",
    "#clusterFilter5 <-table(stdIntegrated$seurat_clusters[clusterFilter5])\n",
    "\n",
    "macroAndMegaClusters <- c(21, 28)\n",
    "DefaultAssay(stdIntegrated) <- \"integrated\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bdb7677a-3030-4a19-b7bd-28e56de5fd65",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Visualize the distribution of cells expressing gene markers across all clusters\n",
    "DefaultAssay(stdIntegrated) <- \"RNA\"\n",
    "FeaturePlot(object = stdIntegrated,\n",
    "            features = c(\"Ms4a3\", \"Clec12a\", \"Fcgr3\"),\n",
    "            reduction = \"tsne\",\n",
    "            label = TRUE,\n",
    "            label.size = 4,\n",
    "            cols = c(\"lightskyblue1\", \"red3\"))\n",
    "# looks like maybe 26\n",
    "VlnPlot(object = stdIntegrated,\n",
    "        features = c(\"Ms4a3\", \"Clec12a\", \"Fcgr3\"),\n",
    "        pt.size = 0.5,\n",
    "        idents = c(26, 0, 1, 2),\n",
    "        ncol = 2)\n",
    "\n",
    "# Determine cutoff values for expression levels of gene markers based on expression across all cells\n",
    "#markerMs4a3 <- stdIntegrated[[\"integrated\"]]@scale.data[\"Ms4a3\",]\n",
    "#quantile(markerMs4a3, probs = seq(0.6, 1, 0.025))\n",
    "\n",
    "#markerClec12a <- stdIntegrated[[\"integrated\"]]@scale.data[\"Clec12a\",]\n",
    "#quantile(markerClec12a, probs = seq(0.6, 1, 0.025))\n",
    "\n",
    "#markerFcgr3 <- stdIntegrated[[\"integrated\"]]@scale.data[\"Fcgr3\",]\n",
    "#quantile(markerFcgr3, probs = seq(0.6, 1, 0.025))\n",
    "\n",
    "# Identify all the cells that satisfy cutoff value which was selected based on distribution\n",
    "#clusterFilter6 <- WhichCells(object = stdIntegrated,\n",
    "#                             expression = Ms4a3 > 0.1 & Clec12a > 0.33 & Fcgr3 > 0.5)\n",
    "\n",
    "# Visualize count of cells per cluster\n",
    "#clusterFilter6 <-table(stdIntegrated$seurat_clusters[clusterFilter6])\n",
    "\n",
    "additionalContaminationCluster <- 26\n",
    "DefaultAssay(stdIntegrated) <- \"integrated\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0449ad33-9b4d-4a5d-a7ee-4ef5130fc50c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae279119-63b1-482b-9433-7fd9ccc69624",
   "metadata": {},
   "outputs": [],
   "source": [
    "bcellClusters <- c(7, 9, 12, 25, 27)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d12afa90-9f8d-4217-80d5-b798c99f1ceb",
   "metadata": {},
   "outputs": [],
   "source": [
    "hematopoieticClusters <- c(bcellClusters,\n",
    "                           erythroidClusters,\n",
    "                           granulocyteClusters,\n",
    "                           macroAndMegaClusters,\n",
    "                           additionalContaminationCluster)\n",
    "\n",
    "# List all clusters that were annotated as non-hematopoietic\n",
    "nonHematopoieticClusters <- as.numeric(levels(stdIntegrated@active.ident)) %>% setdiff(y = hematopoieticClusters)\n",
    "\n",
    "# Total all the cells that were caught by filters per each cluster\n",
    "hematopoieticFilterCount <- c()\n",
    "for(i in 1:length(hematopoieticClusters)) {\n",
    "  hematopoieticFilterCount[i] <- CellsByIdentities(object = stdIntegrated,\n",
    "                                                   idents = hematopoieticClusters[i]) %>%\n",
    "                                   as_vector() %>%\n",
    "                                   length()\n",
    "}\n",
    "sum(hematopoieticFilterCount)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e7d8cf68-e56f-45ce-af2a-cbc1be9cc22b",
   "metadata": {},
   "outputs": [],
   "source": [
    "stdStromal <- subset(x = stdIntegrated,  idents = nonHematopoieticClusters)\n",
    "\n",
    "# Rerun the PCA on the stromal population clusters alone\n",
    "stdStromal <- RunPCA(object = stdStromal, npcs = 50)\n",
    "\n",
    "# Recalculate the neighbors and clusters for the stromal population \n",
    "stdStromal <- FindNeighbors(object = stdStromal, dims = 1:50)\n",
    "\n",
    "stdStromal <- FindClusters(object = stdStromal, resolution = 0.38, random.seed = 2019)\n",
    "\n",
    "# Rerun the tSNE on the stromal population\n",
    "stdStromal <- RunTSNE(object = stdStromal, dims = 1:50)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "88e48400-0c8b-4555-b470-89a95df7bfed",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot the new tSNE of stromal population\n",
    "options(repr.plot.width = 7, repr.plot.height = 5, repr.plot.res = 300)\n",
    "DimPlot( object = stdStromal, reduction = \"tsne\", label = TRUE, label.size = 5) \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2916bd85-8807-4910-8010-b525f6e3b95a",
   "metadata": {},
   "outputs": [],
   "source": [
    "saveRDS(object = stdStromal, file = \"stdStromal.rds\")\n",
    "#stdStromal <- readRDS(file = \"data/stdStromal.rds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba34e280-423a-489f-9ef8-713317b87b8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot keystone gene markers for all 6 major populations\n",
    "options(repr.plot.width = 17, repr.plot.height = 10, repr.plot.res = 300)\n",
    "DefaultAssay(stdStromal) <- \"RNA\"\n",
    "FeaturePlot(object = stdStromal,\n",
    "            features = c(\"Cxcl12\", \"Bglap\", \"Cdh5\", \"Acan\", \"S100a4\", \"Acta2\"),\n",
    "            reduction = \"tsne\", label = TRUE, label.size = 5, ncol = 3, cols = c(\"lightskyblue1\", \"red3\"))\n",
    "DefaultAssay(stdStromal) <- \"integrated\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a17fe0f1-33b4-4079-bdab-463bf06e7e82",
   "metadata": {},
   "outputs": [],
   "source": [
    "atlas = as.data.table(read_excel('~/NIHMS1529101-supplement-8.xlsx', skip = 1))\n",
    "head(atlas)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c243a36-c9b9-4037-a1ec-be0764176b1a",
   "metadata": {},
   "source": [
    "## marker genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f31f68ad-e541-4aaa-8605-e42958eaafce",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Markers for each population as per Scadden et al.\n",
    "marker_MSC = c(\"Lepr\", \"Adipoq\", \"Cxcl12\", \"Kitl\", \"Grem1\", \"Vcam1\")\n",
    "marker_OLC = c(\"Bglap\", \"Runx2\", \"Sp7\", \"Cd200\", \"Spp1\", \"Grem1\", \"Mmp13\")\n",
    "marker_Chondro = c(\"Sox9\", \"Acan\", \"Col2a1\", \"Ihh\", \"Pth1r\", \"Mef2c\")\n",
    "marker_Pericyte = c(\"Acta2\", \"Nes\", \"Cspg4\", \"Myh11\", \"Mcam\")\n",
    "marker_EC = c(\"Pecam1\", \"Cdh5\", \"Cd34\", \"Kdr\", \"Emcn\", \"Flt4\", \"Ly6a\")\n",
    "marker_Fibro = c(\"Fn1\", \"S100a4\", \"Dcn\", \"Sema3c\", \"Cd34\", \"Ly6a\", \"Pdgfra\", \"Thy1\", \"Cd44\", \"Sox9\", \"Scx\", \"Spp1\", \"Nt5e\", \"Cspg4\", \"Cilp\")\n",
    "marker_LeprMSC = c(\"Lepr\", \"Adipoq\", \"Cxcl12\", \"Kitl\", \"Grem1\", \"Vcam1\")\n",
    "marker_OLC1 = c(\"Bglap\", \"Runx2\", \"Sp7\", \"Cxcl12 high\", \"Cd200\", \"Spp1\")\n",
    "marker_OLC2 = c(\"Bglap\", \"Runx2\", \"Sp7\", \"Cxcl12\", \"Grem1\", \"Mmp13\")\n",
    "marker_Chondrohyper = c(\"Sox9\", \"Acan\", \"Col10a1\", \"Ihh\", \"Runx2\", \"Mef2c\")\n",
    "marker_Chondroprogen = c(\"Sox9\", \"Acan\", \"Col2a1\", \"Grem1\", \"Runx2\", \"Sp7\", \"Alpl\", \"Spp1\")\n",
    "marker_Chondro = c(\"Sox9\", \"Acan\", \"Col2a1\")\n",
    "marker_Chondroprehyper2 = c(\"Sox9\", \"Acan\", \"Col2a1\", \"Ihh\", \"Pth1r\", \"Mef2c\")\n",
    "marker_ChondroprolRest = c(\"Sox9\", \"Acan\", \"Col2a1\")\n",
    "marker_Pericytes = c(\"Acta2\", \"Nes\", \"Cspg4\", \"Myh11\", \"Mcam\")\n",
    "marker_ECsinusoidal = c(\"Flt4\", \"Ly6a\", \"Cd34\", \"Il6st\")\n",
    "marker_ECarteriolar = c(\"Flt4\", \"Ly6a\", \"Cd34\", \"Il6st\")\n",
    "marker_ECarterial = c(\"Flt4\", \"Ly6a\", \"Cd34\", \"Il6st\", \"Kitl\")\n",
    "marker_Fibro = c(\"Fn1\", \"S100a4\", \"Dcn\", \"Sema3c\", \"Cd34\", \"Ly6a\", \"Pdgfra\",  \"Thy1\", \"Cd44\", \"Sox9\", \"Scx\", \"Spp1\", \"Nt5e\", \"Cspg4\", \"Cilp\")\n",
    "marker_LeprMSC = c(\"Lepr\", \"Adipoq\", \"Cxcl12\", \"Kitl\", \"Grem1\", \"Vcam1\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b095bb4e-5d7a-47f3-bac2-7e86864aa8aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_markers = unique(c(\"Lepr\", \"Adipoq\", \"Cxcl12\", \"Kitl\", \"Grem1\", \"Vcam1\", \"Bglap\", \"Runx2\", \"Sp7\", \"Cd200\", \"Spp1\", \"Grem1\", \"Mmp13\", \"Sox9\", \"Acan\", \"Col2a1\", \"Ihh\", \"Pth1r\", \"Mef2c\", \"Acta2\", \"Nes\", \"Cspg4\", \"Myh11\", \"Mcam\",\n",
    "  \"Pecam1\", \"Cdh5\", \"Cd34\", \"Kdr\", \"Emcn\", \"Flt4\", \"Ly6a\", \"Lepr\", \"Adipoq\", \"Cxcl12\", \"Kitl\", \"Grem1\", \"Vcam1\", \"Bglap\", \"Runx2\", \"Sp7\", \"Cxcl12\", \"Cd200\", \"Spp1\",\n",
    "  \"Bglap\", \"Runx2\", \"Sp7\", \"Cxcl12\", \"Grem1\", \"Mmp13\", \"Sox9\", \"Acan\", \"Col10a1\", \"Ihh\", \"Runx2\", \"Mef2c\", \"Sox9\", \"Acan\", \"Col2a1\", \"Grem1\", \"Runx2\", \"Sp7\", \"Alpl\", \"Spp1\",\n",
    "  \"Sox9\", \"Acan\", \"Col2a1\", \"Sox9\", \"Acan\", \"Col2a1\", \"Ihh\", \"Pth1r\", \"Mef2c\", \"Sox9\", \"Acan\", \"Col2a1\",\n",
    "  \"Acta2\", \"Nes\", \"Cspg4\", \"Myh11\", \"Mcam\", \"Flt4\", \"Ly6a\", \"Cd34\", \"Il6st\", \"Flt4\", \"Ly6a\", \"Cd34\", \"Il6st\",\n",
    "  \"Flt4\", \"Ly6a\", \"Cd34\", \"Il6st\", \"Kitl\", \"Fn1\", \"S100a4\", \"Dcn\", \"Sema3c\", \"Cd34\", \"Ly6a\", \"Pdgfra\",  \"Thy1\", \"Cd44\", \"Sox9\", \"Scx\", \"Spp1\", \"Nt5e\", \"Cspg4\", \"Cilp\",\n",
    "  \"Fn1\", \"S100a4\", \"Dcn\", \"Sema3c\", \"Cd34\", \"Ly6a\", \"Pdgfra\", \"Thy1\", \"Cd44\", \"Sox9\", \"Scx\", \"Spp1\", \"Nt5e\", \"Cspg4\", \"Cilp\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f216c66b-d505-4150-bb8f-def872cc49d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add interested gene \n",
    "all_markers = c(all_markers, 'S100a6', 'S100a8', 'S100a9')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee1010af-6da6-4d1d-8e64-ab681ffc1016",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 10, repr.plot.height = 6, repr.plot.res = 200)\n",
    "feat_sel = all_markers\n",
    "feat_sel = c(feat_sel, 'LAT2')\n",
    "rna_seu_5_copy = ScaleData(rna_seu_5, features = feat_sel, assay = 'RNA') \n",
    "ret = DoHeatmap(rna_seu_5_copy, features = feat_sel, group.by = 'seurat_clusters_adj', assay = 'RNA', slot = 'scale.data') + \n",
    "scale_fill_gradientn(colors = c(\"blue\", \"white\", \"red\"))\n",
    "ret\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5b1ffe9-7070-426f-9300-0c4ba08f70a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 15, repr.plot.height = 4, repr.plot.res = 200)\n",
    "DotPlot(rna_seu_5_copy, features = feat_sel, group.by = 'seurat_clusters_adj') + \n",
    "theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = .5))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da6d4162-bad2-4e54-9d2f-3ed0e927b7fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "# '~/Copy of Sabatini Metabolic Gene List (Human and Mouse).xlsx'\n",
    "# meta = as.data.table(read_excel('~/Mammalian_Metabolic_Final.xlsx'))\n",
    "# https://esbl.nhlbi.nih.gov/Databases/KSBP2/Targets/Lists/MetabolicEnzymes/MetabolicEnzymeDatabase.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "988314bd-f5b4-454c-a7ec-a38cb87a45c5",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "meta = fread('/research_jude/rgs01_jude/groups/jxugrp/home/common/Lab_Members/WenhuoHu/repos/niche/data/Mouse_metab_genes.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a01cf1d3-f521-45ac-9665-592f57edcc25",
   "metadata": {},
   "outputs": [],
   "source": [
    "head(rna_seu_5_markers[gene %in% meta$Symbol, ])\n",
    "table(rna_seu_5_markers[gene %in% meta$Symbol, cluster])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8fa0e02-737f-4378-af20-fa92f0ea837b",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 13, repr.plot.height = 6, repr.plot.res = 200)\n",
    "#feat_sel = rna_seu_5_markers[gene %in% meta$Symbol,] %>% group_by(cluster) %>% slice_min(p_val_adj, n = 20)\n",
    "#feat_sel = feat_sel$gene\n",
    "rna_seu_5_copy = ScaleData(rna_seu_5, features = feat_sel, assay = 'RNA') \n",
    "ret = DoHeatmap(rna_seu_5_copy, features = feat_sel, group.by = 'seurat_clusters_adj', assay = 'RNA', slot = 'scale.data') + \n",
    "scale_fill_gradientn(colors = c(\"blue\", \"white\", \"red\"))\n",
    "ret\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "efd0c968-9afa-40de-8cda-22a31cfbdae9",
   "metadata": {},
   "outputs": [],
   "source": [
    "feat_sel = rna_seu_5_markers[cluster == '6' & gene %in% meta$Symbol & p_val_adj < 0.05 & avg_log2FC > 1, ]\n",
    "dim(feat_sel)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4984f9d3-99a8-4747-a0d8-b31c653de099",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 15, repr.plot.height = 4, repr.plot.res = 200)\n",
    "feat_sel = rna_seu_5_markers[cluster == '6' & gene %in% meta$Symbol & p_val_adj < 0.05 & avg_log2FC > 1, ][1:60, gene]\n",
    "rna_seu_5_copy = ScaleData(rna_seu_5, features = feat_sel, assay = 'RNA') \n",
    "DotPlot(rna_seu_5_copy, features = feat_sel, group.by = 'seurat_clusters_adj') + \n",
    "theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = .5))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f099f5a-155e-4ee1-99f1-4ba0d447509e",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 15, repr.plot.height = 4, repr.plot.res = 200)\n",
    "feat_sel = rna_seu_5_markers[cluster == '6' & gene %in% meta$Symbol & p_val_adj < 0.05 & avg_log2FC > 1, ][61:120, gene]\n",
    "rna_seu_5_copy = ScaleData(rna_seu_5, features = feat_sel, assay = 'RNA') \n",
    "DotPlot(rna_seu_5_copy, features = feat_sel, group.by = 'seurat_clusters_adj') + \n",
    "theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = .5))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb5ad999-7dce-450a-bf0d-d19dbaff818b",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 15, repr.plot.height = 4, repr.plot.res = 200)\n",
    "feat_sel = rna_seu_5_markers[cluster == '6' & gene %in% meta$Symbol & p_val_adj < 0.05 & avg_log2FC > 1, ][121:171, gene]\n",
    "rna_seu_5_copy = ScaleData(rna_seu_5, features = feat_sel, assay = 'RNA') \n",
    "DotPlot(rna_seu_5_copy, features = feat_sel, group.by = 'seurat_clusters_adj') + \n",
    "theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = .5))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be2257cf-f4c3-439a-bcbc-e504a3092ee8",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 15, repr.plot.height = 4, repr.plot.res = 200)\n",
    "DotPlot(rna_seu_5_copy, features = feat_sel, group.by = 'seurat_clusters_adj') + \n",
    "theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = .5))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f51bb41-a925-4207-a327-e385894d2316",
   "metadata": {},
   "source": [
    "# monocle3 trajectory analysis "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f33882c9-8b3c-400d-865b-32c416fecc64",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(monocle3)\n",
    "library(SeuratWrappers)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca7c31e4-870f-461f-8849-c3a9b7f3145e",
   "metadata": {},
   "outputs": [],
   "source": [
    "GM_state <- function(cds){\n",
    "  if (length(unique(pData(cds)$State)) > 1){\n",
    "    T0_counts <- table(pData(cds)$State, pData(cds)$Hours)[,\"0\"]\n",
    "    return(as.numeric(names(T0_counts)[which\n",
    "          (T0_counts == max(T0_counts))]))\n",
    "  } else {\n",
    "    return (1)\n",
    "  }\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e54d2509-42d6-4872-ba5a-53556b98c0e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_seu_5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0328352-dbe4-4aa3-93b2-cdcdafff2687",
   "metadata": {},
   "outputs": [],
   "source": [
    "cds <- as.cell_data_set(rna_seu_5, )\n",
    "#cds <- preprocess_cds(cds, num_dim = 50)\n",
    "#cds <- reduce_dimension(cds)\n",
    "cds <- cluster_cells(cds, resolution=1e-3)\n",
    "\n",
    "p1 <- plot_cells(cds, color_cells_by = \"cluster\", show_trajectory_graph = FALSE) #, root_state =)\n",
    "p2 <- plot_cells(cds, color_cells_by = \"partition\", show_trajectory_graph = FALSE)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a6be0e5-7fe6-4681-a16e-99f5fae6267c",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 10, repr.plot.height = 5, repr.plot.res = 200)\n",
    "wrap_plots(p1, p2)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f807bf24-3244-4b19-94e1-bcd65f37b327",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a9f483f-e5c7-4067-a8c1-fa19f2605b7c",
   "metadata": {},
   "outputs": [],
   "source": [
    "rowData(cds) = data.frame(gene_short_name = rownames(cds) )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da540fd9-ba43-4ac0-b870-f183e00f59d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "cds <- learn_graph(cds, use_partition = F, verbose = FALSE)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37c29e75-43fb-42f0-b8fa-4a4286233e8b",
   "metadata": {},
   "outputs": [],
   "source": [
    "saveRDS(cds, file = 'data/msc_monocle3.rds')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a1e7327-42e4-407e-9d11-f2b420f84209",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 5, repr.plot.height = 4, repr.plot.res = 200)\n",
    "plot_cells(cds, color_cells_by = \"cluster\", label_groups_by_cluster=FALSE, label_leaves=FALSE, label_branch_points=FALSE)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60320d5d-b87c-42d9-b976-3aa37a999a16",
   "metadata": {},
   "outputs": [],
   "source": [
    "# saved var\n",
    "cds <- order_cells(cds, root_cells = colnames(cds[,cds$seurat_clusters_adj == 1]))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b15a240f-c063-49fb-b3e7-b08535b829fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 5, repr.plot.height = 4, repr.plot.res = 200)\n",
    "plot_cells(cds, \n",
    "           color_cells_by = \"pseudotime\",\n",
    "           group_cells_by = \"cluster\",\n",
    "           label_cell_groups = FALSE,\n",
    "           label_groups_by_cluster=FALSE,\n",
    "           label_leaves=FALSE,\n",
    "           label_branch_points=FALSE,\n",
    "           label_roots = F,\n",
    "           trajectory_graph_color = \"grey60\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08ec2c14-9090-4531-8037-a06c85584a98",
   "metadata": {},
   "outputs": [],
   "source": [
    "marker_test_res <- top_markers(cds, group_cells_by=\"seurat_clusters_adj\",  cores=8)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54509f87-4757-4c2a-9073-8e015bf19039",
   "metadata": {},
   "outputs": [],
   "source": [
    "top_specific_markers <- marker_test_res %>% filter(fraction_expressing >= 0.10) %>% group_by(cell_group) %>% top_n(5, pseudo_R2)\n",
    "\n",
    "top_specific_marker_ids <- unique(top_specific_markers %>% pull(gene_id))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "120f70af-fc62-43a2-8720-502f4743917e",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_cells(cds, genes=c(\"S100a6\"), show_trajectory_graph=FALSE, label_cell_groups=FALSE, label_leaves=FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a887a682-541e-4707-80f5-3271ff89921f",
   "metadata": {},
   "outputs": [],
   "source": [
    "ciliated_cds_pr_test_res <- graph_test(cds, neighbor_graph=\"principal_graph\", cores=4)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ddc24047-e808-4ba4-903b-aed47075570e",
   "metadata": {},
   "outputs": [],
   "source": [
    "pr_deg_ids = as.data.table(ciliated_cds_pr_test_res, keep.rownames = T)\n",
    "pr_deg_ids = pr_deg_ids[order(q_value, -morans_test_statistic), ]\n",
    "head(pr_deg_ids)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ccb6915f-22db-418a-8fc3-538f8aa9ca55",
   "metadata": {},
   "outputs": [],
   "source": [
    "pr_deg_ids <- row.names(subset(ciliated_cds_pr_test_res, q_value < 0.001))\n",
    "length(pr_deg_ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73d9ea8e-d87a-4853-9b63-8c511efbe387",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_module_df <- find_gene_modules(cds[pr_deg_ids,], resolution=c(10^seq(-6,-1)))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "86e93be4-cddb-4841-8ded-35dbc590237f",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_cells(cds, genes=gene_module_df %>% filter(module %in% c(27, 10, 7, 30)),\n",
    "           label_cell_groups=FALSE, show_trajectory_graph=FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "83a1108a-c3b0-41bc-a9fe-caefd37b211e",
   "metadata": {},
   "outputs": [],
   "source": [
    "AFD_genes <- c(\"gcy-8\", \"dac-1\", \"oig-8\")\n",
    "AFD_lineage_cds <- cds[rowData(cds)$gene_short_name %in% AFD_genes,\n",
    "                       colData(cds)$cell.type %in% c(\"AFD\")]\n",
    "AFD_lineage_cds <- order_cells(AFD_lineage_cds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cec4b02b-12b4-4a79-97c8-b9f35eb5a4f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 10, repr.plot.height = 5, repr.plot.res = 200)\n",
    "plot_genes_in_pseudotime(cds[rowData(cds)$gene_short_name %in% c('S100a6', pr_deg_ids[1:5, rn]), ],\n",
    "                         #color_cells_by=\"embryo.time.bin\",\n",
    "                         min_expr=0.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d62ba998-aa53-4975-afc9-ee4e999e34dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(velocyto.R)\n",
    "library(SeuratWrappers)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "356a74ee-4768-4f75-8f9d-5c2d055f6bef",
   "metadata": {},
   "outputs": [],
   "source": [
    "design[, velo_dir := paste0(getwd(), '/velo/', sample_name, '/')]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f50acaf3-8685-4307-8907-c0a5e7a504ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "design[, {dir.create(velo_dir)}, by = 1:nrow(design) ]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2d94c5e-af9f-41ca-86a7-5976bf44a1d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "design[1, velo_dir]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a1c7048-6e1f-4e08-a775-6b61b8b262fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "mm10_gtf = 'research/groups/jxugrp/home/common/Lab_Members/WenhuoHu/igenomes/Mus_musculus/UCSC/mm10/Annotation/Archives/archive-2015-07-17-14-33-26/Genes/genes.gtf'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f725490d-57be-432f-81eb-5eea7557ceb5",
   "metadata": {},
   "outputs": [],
   "source": [
    "design[, gex_bam_file := sub('atac_', 'gex_', bam_file) ]\n",
    "design[1, gex_bam_file]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c115ef6a-108d-41c2-95b6-3b2d0c35ad13",
   "metadata": {},
   "outputs": [],
   "source": [
    "file.exists(design$gex_bam_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a80f86a1-f82a-4787-93f0-d55b888bee91",
   "metadata": {},
   "outputs": [],
   "source": [
    "GTF_FILE=\"/research/groups/jxugrp/home/common/Lab_Members/SyedAhmad/Projects_Main_Farhan/trimmed_copy_leukemia_samples/testig_new_T2T_GTF/hs1.ncbiRefSeq.gtf\"\n",
    "TE_GTF_FILE=\"/research/groups/jxugrp/home/common/Lab_Members/SyedAhmad/Projects_Main_Farhan/trimmed_copy_leukemia_samples/testig_new_T2T_GTF/T2T_CHM13_v2_rmsk_TE.gtf\"\n",
    "bin_tecount = '/home/whu78/conda3/envs/te/bin/'\n",
    "\n",
    "mm10_gtf = '/research/groups/jxugrp/home/common/Lab_Members/WenhuoHu/igenomes/Mus_musculus/UCSC/mm10/Annotation/Archives/archive-2015-07-17-14-33-26/Genes/genes.gtf'\n",
    "design[, gex_bam_file := sub('atac_', 'gex_', bam_file) ]\n",
    "\n",
    "design[, velo_jobname := paste0('velo_', sample_name)]\n",
    "design[, velo_oo := paste0(velo_dir, 'velo_', sample_name, '_out')]\n",
    "design[, velo_eo := paste0(velo_dir, 'velo_', sample_name, '_err')]\n",
    "\n",
    "design[, velo_cmd := paste0('bsub -J ', velo_jobname, ' -n 10 -o ', velo_oo, ' -e ', velo_eo, ' -P velo -R \"rusage[mem=5GB]\" ')]\n",
    "design[, velo_cmd := paste0(velo_cmd, ' -W 666:00')]\n",
    "design[, velo_cmd := paste0(velo_cmd, ' \" velocyto run -@ 10 -o ', velo_dir, ' ', design$gex_bam_file, ' ', mm10_gtf, '\"')]\n",
    "\n",
    "write(design[, velo_cmd], file = paste0(base_dir, 'run_velo.sh'))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51977be1-1cca-4ca7-a380-930f34a47ac8",
   "metadata": {},
   "outputs": [],
   "source": [
    "tmp = fread('velo/loom_list', header = F)\n",
    "setnames(tmp, 1, 'loom_file')\n",
    "tmp[, sample_name := sub(\"..(.*).gex.*\", '\\\\1', loom_file) ]\n",
    "tmp[, loom_file := paste0('/research_jude/rgs01_jude/groups/jxugrp/home/common/Lab_Members/WenhuoHu/collab_JinWang/velo/', loom_file) ]\n",
    "tmp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92e86479-adf4-4eaa-b77c-90294d2d6804",
   "metadata": {},
   "outputs": [],
   "source": [
    "design = merge(design, tmp, by = 'sample_name')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "567a2079-6539-4de4-95dc-9955605c55f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "ma9_msc_samples = c('M2', 'M3', 'M4', 'M4_m10', 'ctrl_m4w', 'ctrl_m6w', 'ctrl_m8w')\n",
    "dsn_sel = design[sample_name %in% ma9_msc_samples,] \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1fcd11e-8e78-43ef-b164-fc6524cdf1b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "msc_design$sample_name\n",
    "dsn_sel$sample_name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "50061956-a35e-47ae-9996-c6cd43429212",
   "metadata": {},
   "outputs": [],
   "source": [
    "mtx_list = lapply(dsn_sel$loom_file, function(fname){ReadVelocity(fname)})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c997ad4-2906-447c-bba7-40e68e204c70",
   "metadata": {},
   "outputs": [],
   "source": [
    "dsn_sel$sampleID2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "871fe0cf-c338-4beb-b00a-9c3154d2fb7c",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_cells = sub('-.*', '', Cells(rna_seu_5))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba756916-a098-4cc1-bb04-5e1833b21543",
   "metadata": {},
   "outputs": [],
   "source": [
    "head(rna_cells)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "114095ba-2820-4ff4-bfae-7532ff3de49d",
   "metadata": {},
   "outputs": [],
   "source": [
    "rowname_list = lapply(1:length(mtx_list), function(ii){\n",
    "    rownames(mtx_list[[ii]][[1]])\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8dff3fb4-b395-40fd-8927-cfc550e867df",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_sel = Reduce(intersect, rowname_list)\n",
    "length(gene_sel) \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0f8b726-fb63-43d4-acb2-3a66426b4172",
   "metadata": {},
   "outputs": [],
   "source": [
    "mtx_list_2 = lapply(1:length(mtx_list), function(ii){\n",
    "    mtx = mtx_list[[ii]]\n",
    "    sname = dsn_sel[ii, sampleID2]\n",
    "    cell_id = colnames(mtx[[1]])\n",
    "    cell_id = sub('.*:', '', cell_id)\n",
    "    cell_id = paste0(sname, '__', cell_id)\n",
    "    \n",
    "    mtx1 = mtx[[1]]\n",
    "    colnames(mtx1) = cell_id\n",
    "    mtx[[1]] = mtx1[gene_sel, cell_id %in% rna_cells]\n",
    "    \n",
    "    mtx2 = mtx[[2]]\n",
    "    colnames(mtx2) = cell_id\n",
    "    mtx[[2]] = mtx2[gene_sel, cell_id %in% rna_cells]\n",
    "    \n",
    "    mtx3 = mtx[[3]]\n",
    "    colnames(mtx3) = cell_id\n",
    "    mtx[[3]] = mtx3[gene_sel, cell_id %in% rna_cells]\n",
    "    mtx\n",
    "    \n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e197c207-4a8c-4f33-a0a2-98769b7143b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "lapply(1:length(mtx_list), function(ii){\n",
    "    dim(mtx_list[[ii]][[1]])\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f36a0375-310f-48df-aa81-0313e574ed42",
   "metadata": {},
   "outputs": [],
   "source": [
    "lapply(1:length(mtx_list_2), function(ii){\n",
    "    dim(mtx_list_2[[ii]][[1]])\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d046f250-ebac-4664-bd78-51d46b940534",
   "metadata": {},
   "outputs": [],
   "source": [
    "mtx1_list = lapply(mtx_list_2, function(xx){xx[[1]]})\n",
    "mtx2_list = lapply(mtx_list_2, function(xx){xx[[2]]})\n",
    "mtx3_list = lapply(mtx_list_2, function(xx){xx[[3]]})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d2e2a92-54ff-497d-8dc7-5402cdc2280c",
   "metadata": {},
   "outputs": [],
   "source": [
    "mtx1 = Reduce(cbind, mtx1_list)\n",
    "mtx2 = Reduce(cbind, mtx2_list)\n",
    "mtx3 = Reduce(cbind, mtx3_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5347d91b-0890-4883-9f89-861cacfd8da2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# saved var\n",
    "mtx_sel_list = list(mtx1, mtx2, mtx3)\n",
    "names(mtx_sel_list) = names(mtx_list[[1]])\n",
    "names(mtx_sel_list)\n",
    "dim(mtx_sel_list[[1]])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fecb62f5-1ad7-4ce9-b023-5558e3f99bdb",
   "metadata": {},
   "outputs": [],
   "source": [
    "tmp = colnames(mtx_sel_list[[1]])\n",
    "tmp2 = make.unique(tmp, sep = '_')\n",
    "colnames(mtx_sel_list[[1]]) = tmp2\n",
    "colnames(mtx_sel_list[[2]]) = tmp2\n",
    "colnames(mtx_sel_list[[3]]) = tmp2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5cee333-6b0e-4fcc-ab59-5674ccab50ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "names(mtx_sel_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61dee03e-304e-4014-a415-bb1a1cf618e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "row_sum = rowSums(mtx_sel_list[[1]])\n",
    "col_sum = colSums(mtx_sel_list[[1]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69038f67-66f5-40b5-9866-29bb15b2e623",
   "metadata": {},
   "outputs": [],
   "source": [
    "sum(row_sum > 1000)\n",
    "sum(col_sum > 1000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df7b2c41-7442-4b33-a387-ea90d58cf1f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "col_sel = col_sum > 1000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16b74596-4205-41dd-8b6e-8b3b18a32564",
   "metadata": {},
   "outputs": [],
   "source": [
    "mtx_sel_list_2 = lapply(mtx_sel_list, function(xx){xx[, col_sel]})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa5b93f8-e88b-4234-8c77-c5152596ec4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "dim(mtx_sel_list_2[[1]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5826aaf-54f6-471d-8481-316e12dc7edd",
   "metadata": {},
   "outputs": [],
   "source": [
    "velo_seu = as.Seurat(mtx_sel_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b0291da-134a-4cd0-b07b-f6effa350393",
   "metadata": {},
   "outputs": [],
   "source": [
    "save.image(file = 'data/nb_niche_MSC_ATAC_R_Jul18.rdata')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2225c8bc-b6e4-4c7e-b432-a3b5785f9397",
   "metadata": {},
   "outputs": [],
   "source": [
    "base::load('data/nb_niche_MSC_ATAC_R_Jul18.rdata')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87941a25-0c58-4931-84a4-03530812c080",
   "metadata": {},
   "outputs": [],
   "source": [
    "velo_seu <- SCTransform(object = velo_seu, assay = \"spliced\")\n",
    "velo_seu <- RunPCA(object = velo_seu, verbose = FALSE)\n",
    "velo_seu <- FindNeighbors(object = velo_seu, dims = 1:20)\n",
    "velo_seu <- FindClusters(object = velo_seu)\n",
    "velo_seu <- RunUMAP(object = velo_seu, dims = 1:20)\n",
    "velo_seu <- RunVelocity(object = velo_seu, deltaT = 1, kCells = 25, fit.quantile = 0.02)\n",
    "ident.colors <- (scales::hue_pal())(n = length(x = levels(x = velo_seu)))\n",
    "names(x = ident.colors) <- levels(x = velo_seu)\n",
    "cell.colors <- ident.colors[Idents(object = velo_seu)]\n",
    "names(x = cell.colors) <- colnames(x = velo_seu)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b0cee1bc-f967-471e-bdb5-7e84fd790eff",
   "metadata": {},
   "outputs": [],
   "source": [
    "velo_seu"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68fabb13-cf12-45f9-b320-ee85c086b5db",
   "metadata": {},
   "outputs": [],
   "source": [
    "getwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29ff8d00-8228-448b-859a-d5e2bd1f1786",
   "metadata": {},
   "outputs": [],
   "source": [
    "velo_seu_tool = Tool(object = velo_seu,  slot = \"RunVelocity\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c646bc2c-94d7-4161-aa44-bd6072fe59a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "names(velo_seu_tool)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "846cd0fc-6c1e-45cb-80e4-c06780abea23",
   "metadata": {},
   "outputs": [],
   "source": [
    "saveRDS(velo_seu_tool, file = 'data/velo_seu_tool.rds')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b64e9f44-b797-4318-a8de-7aa8b30943f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "saveRDS(velo_seu, file = 'data/velo_seu.rds')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "709930e3-01bd-4ab1-b8b8-9d2e987a1e08",
   "metadata": {},
   "outputs": [],
   "source": [
    "show.velocity.on.embedding.cor(emb = Embeddings(object = velo_seu, reduction = \"umap\"), vel = velo_seu_tool, n = 200, \n",
    "                               scale = \"sqrt\", cell.colors = ac(x = cell.colors, alpha = 0.5), \n",
    "                               cex = 0.8, arrow.scale = 3, show.grid.flow = TRUE, min.grid.cell.mass = 0.5, grid.n = 40, arrow.lwd = 1, \n",
    "                               do.par = FALSE, cell.border.alpha = 0.1)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "241f2a03-2f9b-41ef-84ea-a2ceb9571c4c",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "source": [
    "# motif analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "687b1639-183f-47a3-a8a5-d158cadd7678",
   "metadata": {},
   "outputs": [],
   "source": [
    "#library(JASPAR2020)\n",
    "library(BSgenome.Mmusculus.UCSC.mm10)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b614c0ed-c666-4ed0-a3c5-0ec2f792a813",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(Signac)\n",
    "library(Seurat)\n",
    "#library(TFBSTools)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48fbdaff-f029-498f-8a64-8e553cb13894",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get a list of motif position frequency matrices from the JASPAR database\n",
    "pfm <- getMatrixSet(\n",
    "  x = JASPAR2020,\n",
    "  opts = list(collection = \"CORE\", tax_group = 'vertebrates', all_versions = FALSE)\n",
    ")\n"
   ]
  },
  {
   "cell_type": "raw",
   "id": "a341b14c-5eff-4b54-b29b-d4f320b19f64",
   "metadata": {},
   "source": [
    "saveRDS(pfm, file = 'data/pfm.rds')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7857da4-3608-4d1c-b5cc-d7e95747c219",
   "metadata": {},
   "outputs": [],
   "source": [
    "pfm = readRDS('data/pfm.rds')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b68e8f6f-0790-4bb2-87a7-d874f50244b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add motif information\n",
    "oo_msc_5 <- AddMotifs(\n",
    "  object = oo_msc_5,\n",
    "  genome = BSgenome.Mmusculus.UCSC.mm10,\n",
    "  pfm = pfm\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "877a1a19-1fcd-4b88-bbcb-f03c3dae875d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# saved var\n",
    "oo_msc_5 = FRiP(oo_msc_5, assay = 'ATAC', total.fragments = 'fragments')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b211702b-eabe-4530-8d6a-4e633f852acf",
   "metadata": {},
   "outputs": [],
   "source": [
    "Idents(oo_msc_5) = oo_msc_5$seurat_clusters_adj\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0656863f-9608-459a-946f-2534e113c2ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "da_peaks <- FindMarkers( object = oo_msc_5, ident.1 = '6',  ident.2 = c('1', '2', '3', '4'),  only.pos = TRUE, test.use = 'LR', min.pct = 0.05, latent.vars = 'nCount_ATAC' )\n",
    "da_peaks <- rownames(da_peaks[da_peaks$p_val < 0.005 & da_peaks$pct.1 > 0.2, ])\n",
    "enriched_motifs  <- FindMotifs( object = oo_msc_5, features = da_peaks )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d4bbdda-1c49-4d98-8ab5-8ecfcf3babd9",
   "metadata": {},
   "outputs": [],
   "source": [
    "head(enriched_motifs)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ace332b-661a-4f90-bc11-c34149da572f",
   "metadata": {},
   "outputs": [],
   "source": [
    "mm = intersect(toupper(unlist(unname(rna_seu_5_markers[cluster == '4', 'gene']))), enriched_motifs[enriched_motifs$p.adjust < 0.05, 'motif.name'])\n",
    "mm_hsa = c('CEBPD', 'CEBPE', 'STAT3', 'JUND', 'KLF5', 'FOSL1')\n",
    "mm_mmu = c('Cebpd', 'Cebpe', 'Stat3', 'Jund', 'Klf5', 'Fosl1')\n",
    "mm_mmu_2 = c('Cebpa', 'Cebpb', 'Cebpd', 'Cebpe', 'Stat3', 'Jund', 'Klf5', 'Fosl1')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "64b661eb-e083-49d4-9f46-73283834880c",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 12, repr.plot.height = 5, repr.plot.res = 300)\n",
    "motif_sel = c('FOSL2', 'FOS', 'JUND', 'JUNB', 'BATF3', 'BACH2', 'NFE2', 'JDP2')\n",
    "MotifPlot( object = oo_msc_5, motifs = toupper(motif_sel), ncol = 4)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38a76320-aec7-482a-b6ff-ae20144cb0cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 12, repr.plot.height = 4, repr.plot.res = 300)\n",
    "motif_sel = c('SMAD3', 'RBPJ', 'RUNX1', 'SP1')\n",
    "MotifPlot( object = oo_msc_5, motifs = toupper(motif_sel), ncol = 4)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f0dc4450-a1c6-479b-b125-0ad557ec8552",
   "metadata": {},
   "outputs": [],
   "source": [
    "motif_sel = c('FOSL2', 'FOS', 'JUND', 'JUNB', 'BATF3', 'BACH2', 'NFE2', 'JDP2', 'SMAD3', 'RBPJ', 'RUNX1', 'SP1')\n",
    "motif_sel = c('Fosl2', 'Fos', 'Jund', 'Junb', 'Batf3', 'Bach2', 'Nfe2', 'Jdp2', 'Smad3', 'Rbpj', 'RUNX1', 'Sp1')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ea60b72-7946-4ed2-9f3e-c095ea5f5040",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 8, repr.plot.height = 3, repr.plot.res = 200)\n",
    "rna_seu_5_copy = ScaleData(rna_seu_5, features = motif_sel, assay = 'RNA') \n",
    "DotPlot(rna_seu_5_copy, features = motif_sel, group.by = 'seurat_clusters_adj') + \n",
    "theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = .5))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d9af5943-3660-4142-8e8b-69bbb6fd8758",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 10, repr.plot.height = 3, repr.plot.res = 200)\n",
    "rna_seu_5_copy = ScaleData(rna_seu_5, features = mm_mmu_2, assay = 'RNA') \n",
    "ret = DoHeatmap(rna_seu_5_copy, features = mm_mmu_2, group.by = 'seurat_clusters_adj', assay = 'RNA', slot = 'scale.data') + \n",
    "scale_fill_gradientn(colors = c(\"blue\", \"white\", \"red\"))\n",
    "ret\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85212ec7-ed08-4b21-ad29-3f575877bcac",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 12, repr.plot.height = 6, repr.plot.res = 300)\n",
    "VlnPlot(object = rna_seu_5, features = mm_mmu_2, ncol = 4, group.by = 'seurat_clusters_adj')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d89fa6f-f805-456b-9ee2-17089f4a740e",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 12, repr.plot.height = 6, repr.plot.res = 300)\n",
    "MotifPlot( object = oo_msc_5, motifs = head(rownames(enriched_motifs), 20) )\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8431188d-9a5a-4074-ab6b-141462e4ad2f",
   "metadata": {},
   "source": [
    "# pathway analysis "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35f1f7a0-cee8-46da-82f2-db25c010e72e",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 5, repr.plot.height = 4, repr.plot.res = 200)\n",
    "DimPlot(rna_seu_5, group.by = c(\"seurat_clusters_adj\"), label = TRUE) + ggtitle(\"RNA\") \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e50f2cec-ab92-4d2b-bfd2-1b18cac1aa5d",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 6, repr.plot.height = 3, repr.plot.res = 200)\n",
    "FeaturePlot(rna_seu_5, features = c('Eef1a1', 'Fth1'), label = TRUE) \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "259dd2b6-55fc-4f59-a46e-9fa4137ee027",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(enrichR)\n",
    "dbname = as.data.table(listEnrichrDbs())\n",
    "dbname[grep('reactome', libraryName, ignore.case = T), ]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7067f283-5eb2-44cb-bc78-189f67684452",
   "metadata": {},
   "outputs": [],
   "source": [
    "dbname[grep('metab', libraryName, ignore.case = T), ]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e43f7f5-221e-4f03-899d-2707b7184611",
   "metadata": {},
   "outputs": [],
   "source": [
    "dbname[grep('Transcription_Factor_PPIs', libraryName, ignore.case = T), ]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11400d79-8c81-4b89-98bc-1e57dce7f496",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_seu_5@active.ident = rna_seu_5$seurat_clusters_adj"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5888540-3857-41c8-bb6c-b00c7218d936",
   "metadata": {},
   "outputs": [],
   "source": [
    "#PPI_Hub_Proteins GeneSigDB KEGG_2013 Reactome_2013 GO_Cellular_Component_2015 GO_Biological_Process_2015 CellMarker_2024\n",
    "options(repr.plot.width = 10, repr.plot.height = 6, repr.plot.res = 300)\n",
    "path_list = DEenrichRPlot(rna_seu_5, ident.1 = c('6'), ident.2 = c('1', '2', '3', '4', '5'), balanced = F, return.gene.list = T, p.val.cutoff = 0.05, enrich.database = \"HMDB_Metabolites\", max.genes = 250, num.pathway = 20)\n",
    "path_list = as.data.table(path_list$pos)\n",
    "path_list = path_list[HMDB_Metabolites.Adjusted.P.value < 0.05, ]\n",
    "path_list\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f3aee74-3717-46f8-bdec-4f2bfc87c0fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "path_list = as.data.table(path_list$pos)\n",
    "path_list = path_list[HMDB_Metabolites.Adjusted.P.value < 0.05, ]\n",
    "path_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "415b191e-f150-4edf-be8c-0bd842d69d82",
   "metadata": {},
   "outputs": [],
   "source": [
    "#PPI_Hub_Proteins GeneSigDB KEGG_2013 Reactome_2013 GO_Cellular_Component_2015 GO_Biological_Process_2015 CellMarker_2024\n",
    "options(repr.plot.width = 10, repr.plot.height = 4, repr.plot.res = 300)\n",
    "up_6vs1 = DEenrichRPlot(rna_seu_5, ident.1 = c('6'), ident.2 = c('1', '2', '3', '4', '5'), return.gene.list = T, balanced = F, enrich.database = \"HMDB_Metabolites\", max.genes = 50, num.pathway = 20, logfc.threshold = 1, p.val.cutoff = 0.05)\n",
    "DEenrichRPlot(rna_seu_5, ident.1 = c('6'), ident.2 = c('1', '2', '3', '4', '5'), balanced = F, enrich.database = \"HMDB_Metabolites\", max.genes = 50, num.pathway = 20, logfc.threshold = 1, p.val.cutoff = 0.05)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35e8c32e-cef1-4251-a6ee-95a162f4f93f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#PPI_Hub_Proteins GeneSigDB KEGG_2013 Reactome_2013 GO_Cellular_Component_2015 GO_Biological_Process_2015 CellMarker_2024\n",
    "options(repr.plot.width = 10, repr.plot.height = 5.5, repr.plot.res = 300)\n",
    "dn_6vs1 = DEenrichRPlot(rna_seu_5, ident.2 = c('6'), ident.1 = c('1', '2', '3', '4', '5'), return.gene.list = T, balanced = F, enrich.database = \"HMDB_Metabolites\", max.genes = 50, num.pathway = 20, logfc.threshold = 1, p.val.cutoff = 0.05)\n",
    "DEenrichRPlot(rna_seu_5, ident.2 = c('6'), ident.1 = c('1', '2', '3', '4', '5'), balanced = F, enrich.database = \"HMDB_Metabolites\", max.genes = 50, num.pathway = 20, logfc.threshold = 1, p.val.cutoff = 0.05)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ed65cd6-cc3d-4696-a40a-2dfa07e6219b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#PPI_Hub_Proteins GeneSigDB KEGG_2013 Reactome_2013 GO_Cellular_Component_2015 GO_Biological_Process_2015 CellMarker_2024\n",
    "options(repr.plot.width = 10, repr.plot.height = 4, repr.plot.res = 300)\n",
    "up_6vs7 = DEenrichRPlot(rna_seu_5, ident.1 = c('6'), ident.2 = c('7'), return.gene.list = T, balanced = F, enrich.database = \"HMDB_Metabolites\", max.genes = 50, num.pathway = 20, logfc.threshold = 1, p.val.cutoff = 0.05)\n",
    "DEenrichRPlot(rna_seu_5, ident.1 = c('6'), ident.2 = c('7'), balanced = F, enrich.database = \"HMDB_Metabolites\", max.genes = 50, num.pathway = 20, logfc.threshold = 1, p.val.cutoff = 0.05)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa63f41b-6193-4f55-997e-3c8d50b4fd20",
   "metadata": {},
   "outputs": [],
   "source": [
    "#PPI_Hub_Proteins GeneSigDB KEGG_2013 Reactome_2013 GO_Cellular_Component_2015 GO_Biological_Process_2015 CellMarker_2024\n",
    "options(repr.plot.width = 10, repr.plot.height = 4, repr.plot.res = 300)\n",
    "dn_6vs7 = DEenrichRPlot(rna_seu_5, ident.2 = c('6'), ident.1 = c('7'), return.gene.list = T, balanced = F, enrich.database = \"HMDB_Metabolites\", max.genes = 50, num.pathway = 20, logfc.threshold = 1, p.val.cutoff = 0.05)\n",
    "DEenrichRPlot(rna_seu_5, ident.2 = c('6'), ident.1 = c('7'), balanced = F, enrich.database = \"HMDB_Metabolites\", max.genes = 50, num.pathway = 20, logfc.threshold = 1, p.val.cutoff = 0.05)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ddfedb70-044c-40e2-9514-9756582071ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "#PPI_Hub_Proteins GeneSigDB KEGG_2013 Reactome_2013 GO_Cellular_Component_2015 GO_Biological_Process_2015 CellMarker_2024\n",
    "options(repr.plot.width = 10, repr.plot.height = 3, repr.plot.res = 300)\n",
    "up_6vsall = DEenrichRPlot(rna_seu_5, ident.1 = c('6'), ident.2 = c('1', '2', '3', '4', '5', '7'), return.gene.list = T, balanced = F, enrich.database = \"HMDB_Metabolites\", max.genes = 50, num.pathway = 20, logfc.threshold = 1, p.val.cutoff = 0.05)\n",
    "DEenrichRPlot(rna_seu_5, ident.1 = c('6'), ident.2 = c('1', '2', '3', '4', '5', '7'), balanced = F, enrich.database = \"HMDB_Metabolites\", max.genes = 50, num.pathway = 20, logfc.threshold = 1, p.val.cutoff = 0.05)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b04d39bd-681e-4ec3-ba0d-d669eacad659",
   "metadata": {},
   "outputs": [],
   "source": [
    "#PPI_Hub_Proteins GeneSigDB KEGG_2013 Reactome_2013 GO_Cellular_Component_2015 GO_Biological_Process_2015 CellMarker_2024\n",
    "options(repr.plot.width = 10, repr.plot.height = 5, repr.plot.res = 300)\n",
    "dn_6vsall = DEenrichRPlot(rna_seu_5, ident.2 = c('6'), ident.1 = c('1', '2', '3', '4', '5', '7'), return.gene.list = T, balanced = F, enrich.database = \"HMDB_Metabolites\", max.genes = 50, num.pathway = 20, logfc.threshold = 1, p.val.cutoff = 0.05)\n",
    "DEenrichRPlot(rna_seu_5, ident.2 = c('6'), ident.1 = c('1', '2', '3', '4', '5', '7'), balanced = F, enrich.database = \"HMDB_Metabolites\", max.genes = 50, num.pathway = 20, logfc.threshold = 1, p.val.cutoff = 0.05)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff46e512-7b0f-45c7-970f-29afd0e1a83a",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 5, repr.plot.height = 3, repr.plot.res = 200)\n",
    "feat = str_to_title(c('NDUFA4L2', 'GAPDH', 'EEF1A1'))\n",
    "DotPlot(rna_seu_5, features = feat)\n",
    "#FeaturePlot(rna_seu_5, features = feat, label = TRUE) \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "857d6f35-b46a-403a-8c1b-f29f55beb6f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 5, repr.plot.height = 3, repr.plot.res = 200)\n",
    "feat = str_to_title(c('LARS2', 'UBR5', 'HUWE1', 'MID1', 'HEXB'))\n",
    "DotPlot(rna_seu_5, features = feat)\n",
    "#FeaturePlot(rna_seu_5, features = feat, label = TRUE) \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28016c04-b91d-497d-b966-a2e438d7d295",
   "metadata": {},
   "outputs": [],
   "source": [
    "#PPI_Hub_Proteins GeneSigDB KEGG_2013 Reactome_2013 GO_Cellular_Component_2015 GO_Biological_Process_2015 CellMarker_2024\n",
    "options(repr.plot.width = 10, repr.plot.height = 6, repr.plot.res = 300)\n",
    "DEenrichRPlot(rna_seu_5, ident.1 = c('5', '6'), ident.2 = c('7'), balanced = F, enrich.database = \"HMDB_Metabolites\", max.genes = 250, num.pathway = 20, logfc.threshold = 1, p.val.cutoff = 0.05)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5ee2841-9de5-4fc1-ab1d-9cfe8d3b803b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#PPI_Hub_Proteins GeneSigDB KEGG_2013 Reactome_2013 GO_Cellular_Component_2015 GO_Biological_Process_2015 CellMarker_2024\n",
    "options(repr.plot.width = 10, repr.plot.height = 3, repr.plot.res = 300)\n",
    "gene_list = DEenrichRPlot(rna_seu_5, ident.1 = c('6'), ident.2 = c('1', '2', '3', '4'), balanced = F, enrich.database = \"Transcription_Factor_PPIs\", return.gene.list = T,\n",
    "              p.val.cutoff = 0.01,  max.genes = 50)\n",
    "# DEenrichRPlot(object = ctrl.test, ident.1 = \"Basal_1\",ident.2 = \"Basal_2\",enrich.database = \"BP\",max.genes = 50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eaeb0b36-eb37-46d4-8ccd-d31abdeedd14",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_list = as.data.table(gene_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b3326f7a-ebdb-4c8f-a732-f5c82e29a440",
   "metadata": {},
   "outputs": [],
   "source": [
    "feat_sel = c('GAPDH', 'RPL5', 'RPL32', 'RPL10', 'RPL11', 'RPL10A', 'RPL8', 'RPL9', 'RPS4X', 'RPS16', 'RPL18A', 'RPS3', 'RPL13', 'RPS2', 'RPL15', 'RPL18', 'RPL19', 'RPL21', 'RPS8', 'RPL23', 'RPL23A', 'EEF1A1', 'RPS28', 'RPL27', 'RPS20', 'RPS23')\n",
    "feat_sel = geneId[feat_sel, Mouse.gene.name]\n",
    "feat_sel = c(feat_sel[feat_sel != ''], 'Gapdh')\n",
    "feat_sel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b0becf56-463a-4204-8014-e14c221d73df",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 7, repr.plot.height = 3, repr.plot.res = 200)\n",
    "rna_seu_5_copy = ScaleData(rna_seu_5, features = feat_sel , assay = 'RNA') \n",
    "ret = DoHeatmap(rna_seu_5_copy, features = feat_sel, group.by = 'seurat_clusters_adj', assay = 'RNA', slot = 'scale.data') + \n",
    "scale_fill_gradientn(colors = c(\"blue\", \"white\", \"red\"))\n",
    "ret\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bbc9e7a4-67af-4482-9fb1-d9029f9ade51",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 7, repr.plot.height = 3, repr.plot.res = 200)\n",
    "DotPlot(rna_seu_5_copy, features = feat_sel, group.by = 'seurat_clusters_adj') + \n",
    "theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = .5))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6c6ae07-1987-4576-a2db-7baadcb6f601",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4dda58f7-5518-4f49-9df6-bd563fa5766e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "mar25R",
   "language": "R",
   "name": "mar25r"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
